<!doctype html>
<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1.0"><title>cxa_exception.cpp source code [codebrowser/external/libcxxabi/src/cxa_exception.cpp] - Woboq Code Browser</title>
<link rel="stylesheet" href="../../../../../data/qtcreator.css" title="QtCreator"/>
<link rel="alternate stylesheet" href="../../../../../data/kdevelop.css" title="KDevelop"/>
<script type="text/javascript" src="../../../../../data/jquery/jquery.min.js"></script>
<script type="text/javascript" src="../../../../../data/jquery/jquery-ui.min.js"></script>
<script>var file = 'codebrowser/external/libcxxabi/src/cxa_exception.cpp'; var root_path = '../../../..'; var data_path = '../../../../../data';</script>
<script src='../../../../../data/codebrowser.js'></script>
</head>
<body><div id='header'><h1 id='breadcrumb'><span>Browse the source code of </span><a href='../../..'>codebrowser</a>/<a href='../..'>external</a>/<a href='..'>libcxxabi</a>/<a href='./'>src</a>/<a href='cxa_exception.cpp.html'>cxa_exception.cpp</a></h1></div>
<hr/><div id='content'><table class="code">
<tr><th id="1">1</th><td><i>//===------------------------- cxa_exception.cpp --------------------------===//</i></td></tr>
<tr><th id="2">2</th><td><i>//</i></td></tr>
<tr><th id="3">3</th><td><i>//                     The LLVM Compiler Infrastructure</i></td></tr>
<tr><th id="4">4</th><td><i>//</i></td></tr>
<tr><th id="5">5</th><td><i>// This file is dual licensed under the MIT and the University of Illinois Open</i></td></tr>
<tr><th id="6">6</th><td><i>// Source Licenses. See LICENSE.TXT for details.</i></td></tr>
<tr><th id="7">7</th><td><i>//</i></td></tr>
<tr><th id="8">8</th><td><i>//  </i></td></tr>
<tr><th id="9">9</th><td><i>//  This file implements the "Exception Handling APIs"</i></td></tr>
<tr><th id="10">10</th><td><i>//  <a href="http://mentorembedded.github.io/cxx-abi/abi-eh.html">http://mentorembedded.github.io/cxx-abi/abi-eh.html</a></i></td></tr>
<tr><th id="11">11</th><td><i>//  </i></td></tr>
<tr><th id="12">12</th><td><i>//===----------------------------------------------------------------------===//</i></td></tr>
<tr><th id="13">13</th><td></td></tr>
<tr><th id="14">14</th><td><u>#include <a href="config.h.html">"config.h"</a></u></td></tr>
<tr><th id="15">15</th><td><u>#include <a href="../include/cxxabi.h.html">"cxxabi.h"</a></u></td></tr>
<tr><th id="16">16</th><td></td></tr>
<tr><th id="17">17</th><td><u>#include <a href="../../libcxx/include/exception.html">&lt;exception&gt;</a>        // for std::terminate</u></td></tr>
<tr><th id="18">18</th><td><u>#include <a href="../../libcxx/include/cstdlib.html">&lt;cstdlib&gt;</a>          // for malloc, free</u></td></tr>
<tr><th id="19">19</th><td><u>#include <a href="../../libcxx/include/cstring.html">&lt;cstring&gt;</a>          // for memset</u></td></tr>
<tr><th id="20">20</th><td><u>#<span data-ppcond="20">if</span> !<a class="macro" href="config.h.html#21" title="0" data-ref="_M/LIBCXXABI_HAS_NO_THREADS">LIBCXXABI_HAS_NO_THREADS</a></u></td></tr>
<tr><th id="21">21</th><td><u>#  include <a href="../../../bionic/libc/include/pthread.h.html">&lt;pthread.h&gt;</a>      // for fallback_malloc.ipp's mutexes</u></td></tr>
<tr><th id="22">22</th><td><u>#<span data-ppcond="20">endif</span></u></td></tr>
<tr><th id="23">23</th><td><u>#include <a href="cxa_exception.hpp.html">"cxa_exception.hpp"</a></u></td></tr>
<tr><th id="24">24</th><td><u>#include <a href="cxa_handlers.hpp.html">"cxa_handlers.hpp"</a></u></td></tr>
<tr><th id="25">25</th><td></td></tr>
<tr><th id="26">26</th><td><i>// +---------------------------+-----------------------------+---------------+</i></td></tr>
<tr><th id="27">27</th><td><i>// | __cxa_exception           | _Unwind_Exception CLNGC++\0 | thrown object |</i></td></tr>
<tr><th id="28">28</th><td><i>// +---------------------------+-----------------------------+---------------+</i></td></tr>
<tr><th id="29">29</th><td><i>//                                                           ^</i></td></tr>
<tr><th id="30">30</th><td><i>//                                                           |</i></td></tr>
<tr><th id="31">31</th><td><i>//   +-------------------------------------------------------+</i></td></tr>
<tr><th id="32">32</th><td><i>//   |</i></td></tr>
<tr><th id="33">33</th><td><i>// +---------------------------+-----------------------------+</i></td></tr>
<tr><th id="34">34</th><td><i>// | __cxa_dependent_exception | _Unwind_Exception CLNGC++\1 |</i></td></tr>
<tr><th id="35">35</th><td><i>// +---------------------------+-----------------------------+</i></td></tr>
<tr><th id="36">36</th><td></td></tr>
<tr><th id="37">37</th><td><b>namespace</b> <span class="namespace">__cxxabiv1</span> {</td></tr>
<tr><th id="38">38</th><td></td></tr>
<tr><th id="39">39</th><td><u>#pragma GCC visibility push(default)</u></td></tr>
<tr><th id="40">40</th><td></td></tr>
<tr><th id="41">41</th><td><i  data-doc="__cxxabiv1::cxa_exception_from_thrown_object">//  Utility routines</i></td></tr>
<tr><th id="42">42</th><td><em>static</em></td></tr>
<tr><th id="43">43</th><td><b>inline</b></td></tr>
<tr><th id="44">44</th><td><a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>*</td></tr>
<tr><th id="45">45</th><td><dfn class="tu decl def" id="__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-type='__cxxabiv1::__cxa_exception * __cxxabiv1::cxa_exception_from_thrown_object(void * thrown_object)' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</dfn>(<em>void</em>* <dfn class="local col1 decl" id="1thrown_object" title='thrown_object' data-type='void *' data-ref="1thrown_object">thrown_object</dfn>)</td></tr>
<tr><th id="46">46</th><td>{</td></tr>
<tr><th id="47">47</th><td>    <b>return</b> <b>static_cast</b>&lt;<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>*&gt;(<a class="local col1 ref" href="#1thrown_object" title='thrown_object' data-ref="1thrown_object">thrown_object</a>) - <var>1</var>;</td></tr>
<tr><th id="48">48</th><td>}</td></tr>
<tr><th id="49">49</th><td></td></tr>
<tr><th id="50">50</th><td><i  data-doc="__cxxabiv1::thrown_object_from_cxa_exception">// Note:  This is never called when exception_header is masquerading as a</i></td></tr>
<tr><th id="51">51</th><td><i  data-doc="__cxxabiv1::thrown_object_from_cxa_exception">//        __cxa_dependent_exception.</i></td></tr>
<tr><th id="52">52</th><td><em>static</em></td></tr>
<tr><th id="53">53</th><td><b>inline</b></td></tr>
<tr><th id="54">54</th><td><em>void</em>*</td></tr>
<tr><th id="55">55</th><td><dfn class="tu decl def" id="__cxxabiv1::thrown_object_from_cxa_exception" title='__cxxabiv1::thrown_object_from_cxa_exception' data-type='void * __cxxabiv1::thrown_object_from_cxa_exception(__cxxabiv1::__cxa_exception * exception_header)' data-ref="__cxxabiv1::thrown_object_from_cxa_exception">thrown_object_from_cxa_exception</dfn>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col2 decl" id="2exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="2exception_header">exception_header</dfn>)</td></tr>
<tr><th id="56">56</th><td>{</td></tr>
<tr><th id="57">57</th><td>    <b>return</b> <b>static_cast</b>&lt;<em>void</em>*&gt;(<a class="local col2 ref" href="#2exception_header" title='exception_header' data-ref="2exception_header">exception_header</a> + <var>1</var>);</td></tr>
<tr><th id="58">58</th><td>}</td></tr>
<tr><th id="59">59</th><td></td></tr>
<tr><th id="60">60</th><td><i  data-doc="__cxxabiv1::cxa_exception_from_exception_unwind_exception">//  Get the exception object from the unwind pointer.</i></td></tr>
<tr><th id="61">61</th><td><i  data-doc="__cxxabiv1::cxa_exception_from_exception_unwind_exception">//  Relies on the structure layout, where the unwind pointer is right in</i></td></tr>
<tr><th id="62">62</th><td><i  data-doc="__cxxabiv1::cxa_exception_from_exception_unwind_exception">//  front of the user's exception object</i></td></tr>
<tr><th id="63">63</th><td><em>static</em></td></tr>
<tr><th id="64">64</th><td><b>inline</b></td></tr>
<tr><th id="65">65</th><td><a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>*</td></tr>
<tr><th id="66">66</th><td><dfn class="tu decl def" id="__cxxabiv1::cxa_exception_from_exception_unwind_exception" title='__cxxabiv1::cxa_exception_from_exception_unwind_exception' data-type='__cxxabiv1::__cxa_exception * __cxxabiv1::cxa_exception_from_exception_unwind_exception(_Unwind_Exception * unwind_exception)' data-ref="__cxxabiv1::cxa_exception_from_exception_unwind_exception">cxa_exception_from_exception_unwind_exception</dfn>(<span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col3 decl" id="3unwind_exception" title='unwind_exception' data-type='_Unwind_Exception *' data-ref="3unwind_exception">unwind_exception</dfn>)</td></tr>
<tr><th id="67">67</th><td>{</td></tr>
<tr><th id="68">68</th><td>    <b>return</b> <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col3 ref" href="#3unwind_exception" title='unwind_exception' data-ref="3unwind_exception">unwind_exception</a> + <var>1</var> );</td></tr>
<tr><th id="69">69</th><td>}</td></tr>
<tr><th id="70">70</th><td></td></tr>
<tr><th id="71">71</th><td><em>static</em></td></tr>
<tr><th id="72">72</th><td><b>inline</b></td></tr>
<tr><th id="73">73</th><td><span class='typedef' title='size_t' data-type='unsigned long' data-ref="size_t">size_t</span></td></tr>
<tr><th id="74">74</th><td><dfn class="tu decl def" id="__cxxabiv1::cxa_exception_size_from_exception_thrown_size" title='__cxxabiv1::cxa_exception_size_from_exception_thrown_size' data-type='size_t __cxxabiv1::cxa_exception_size_from_exception_thrown_size(size_t size)' data-ref="__cxxabiv1::cxa_exception_size_from_exception_thrown_size">cxa_exception_size_from_exception_thrown_size</dfn>(<span class='typedef' title='size_t' data-type='unsigned long' data-ref="size_t">size_t</span> <dfn class="local col4 decl" id="4size" title='size' data-type='size_t' data-ref="4size">size</dfn>)</td></tr>
<tr><th id="75">75</th><td>{</td></tr>
<tr><th id="76">76</th><td>    <b>return</b> <a class="local col4 ref" href="#4size" title='size' data-ref="4size">size</a> + <b>sizeof</b> (<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>);</td></tr>
<tr><th id="77">77</th><td>}</td></tr>
<tr><th id="78">78</th><td></td></tr>
<tr><th id="79">79</th><td><em>static</em> <em>void</em> <dfn class="tu decl def" id="__cxxabiv1::setExceptionClass" title='__cxxabiv1::setExceptionClass' data-type='void __cxxabiv1::setExceptionClass(_Unwind_Exception * unwind_exception)' data-ref="__cxxabiv1::setExceptionClass">setExceptionClass</dfn>(<span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col5 decl" id="5unwind_exception" title='unwind_exception' data-type='_Unwind_Exception *' data-ref="5unwind_exception">unwind_exception</dfn>) {</td></tr>
<tr><th id="80">80</th><td>    <a class="local col5 ref" href="#5unwind_exception" title='unwind_exception' data-ref="5unwind_exception">unwind_exception</a>-&gt;<span class='ref' title='_Unwind_Exception::exception_class' data-ref="_Unwind_Exception::exception_class">exception_class</span> = <a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::kOurExceptionClass" title='__cxxabiv1::kOurExceptionClass' data-ref="__cxxabiv1::kOurExceptionClass">kOurExceptionClass</a>;</td></tr>
<tr><th id="81">81</th><td>}</td></tr>
<tr><th id="82">82</th><td></td></tr>
<tr><th id="83">83</th><td><em>static</em> <em>void</em> <dfn class="tu decl def" id="__cxxabiv1::setDependentExceptionClass" title='__cxxabiv1::setDependentExceptionClass' data-type='void __cxxabiv1::setDependentExceptionClass(_Unwind_Exception * unwind_exception)' data-ref="__cxxabiv1::setDependentExceptionClass">setDependentExceptionClass</dfn>(<span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col6 decl" id="6unwind_exception" title='unwind_exception' data-type='_Unwind_Exception *' data-ref="6unwind_exception">unwind_exception</dfn>) {</td></tr>
<tr><th id="84">84</th><td>    <a class="local col6 ref" href="#6unwind_exception" title='unwind_exception' data-ref="6unwind_exception">unwind_exception</a>-&gt;<span class='ref' title='_Unwind_Exception::exception_class' data-ref="_Unwind_Exception::exception_class">exception_class</span> = <a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::kOurDependentExceptionClass" title='__cxxabiv1::kOurDependentExceptionClass' data-ref="__cxxabiv1::kOurDependentExceptionClass">kOurDependentExceptionClass</a>;</td></tr>
<tr><th id="85">85</th><td>}</td></tr>
<tr><th id="86">86</th><td></td></tr>
<tr><th id="87">87</th><td><i  data-doc="__cxxabiv1::isOurExceptionClass">//  Is it one of ours?</i></td></tr>
<tr><th id="88">88</th><td><em>static</em> <em>bool</em> <dfn class="tu decl def" id="__cxxabiv1::isOurExceptionClass" title='__cxxabiv1::isOurExceptionClass' data-type='bool __cxxabiv1::isOurExceptionClass(const _Unwind_Exception * unwind_exception)' data-ref="__cxxabiv1::isOurExceptionClass">isOurExceptionClass</dfn>(<em>const</em> <span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col7 decl" id="7unwind_exception" title='unwind_exception' data-type='const _Unwind_Exception *' data-ref="7unwind_exception">unwind_exception</dfn>) {</td></tr>
<tr><th id="89">89</th><td>    <b>return</b> (<a class="local col7 ref" href="#7unwind_exception" title='unwind_exception' data-ref="7unwind_exception">unwind_exception</a>-&gt;<span class='ref' title='_Unwind_Exception::exception_class' data-ref="_Unwind_Exception::exception_class">exception_class</span> &amp; <a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::get_vendor_and_language" title='__cxxabiv1::get_vendor_and_language' data-ref="__cxxabiv1::get_vendor_and_language">get_vendor_and_language</a>) == </td></tr>
<tr><th id="90">90</th><td>           (<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::kOurExceptionClass" title='__cxxabiv1::kOurExceptionClass' data-ref="__cxxabiv1::kOurExceptionClass">kOurExceptionClass</a>                &amp; <a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::get_vendor_and_language" title='__cxxabiv1::get_vendor_and_language' data-ref="__cxxabiv1::get_vendor_and_language">get_vendor_and_language</a>);</td></tr>
<tr><th id="91">91</th><td>}</td></tr>
<tr><th id="92">92</th><td></td></tr>
<tr><th id="93">93</th><td><em>static</em> <em>bool</em> <dfn class="tu decl def" id="__cxxabiv1::isDependentException" title='__cxxabiv1::isDependentException' data-type='bool __cxxabiv1::isDependentException(_Unwind_Exception * unwind_exception)' data-ref="__cxxabiv1::isDependentException">isDependentException</dfn>(<span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col8 decl" id="8unwind_exception" title='unwind_exception' data-type='_Unwind_Exception *' data-ref="8unwind_exception">unwind_exception</dfn>) {</td></tr>
<tr><th id="94">94</th><td>    <b>return</b> (<a class="local col8 ref" href="#8unwind_exception" title='unwind_exception' data-ref="8unwind_exception">unwind_exception</a>-&gt;<span class='ref' title='_Unwind_Exception::exception_class' data-ref="_Unwind_Exception::exception_class">exception_class</span> &amp; <var>0xFF</var>) == <var>0x01</var>;</td></tr>
<tr><th id="95">95</th><td>}</td></tr>
<tr><th id="96">96</th><td></td></tr>
<tr><th id="97">97</th><td><i  data-doc="__cxxabiv1::incrementHandlerCount">//  This does not need to be atomic</i></td></tr>
<tr><th id="98">98</th><td><em>static</em> <b>inline</b> <em>int</em> <dfn class="tu decl def" id="__cxxabiv1::incrementHandlerCount" title='__cxxabiv1::incrementHandlerCount' data-type='int __cxxabiv1::incrementHandlerCount(__cxxabiv1::__cxa_exception * exception)' data-ref="__cxxabiv1::incrementHandlerCount">incrementHandlerCount</dfn>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a> *<dfn class="local col9 decl" id="9exception" title='exception' data-type='__cxxabiv1::__cxa_exception *' data-ref="9exception">exception</dfn>) {</td></tr>
<tr><th id="99">99</th><td>    <b>return</b> ++<a class="local col9 ref" href="#9exception" title='exception' data-ref="9exception">exception</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a>;</td></tr>
<tr><th id="100">100</th><td>}</td></tr>
<tr><th id="101">101</th><td></td></tr>
<tr><th id="102">102</th><td><i  data-doc="__cxxabiv1::decrementHandlerCount">//  This does not need to be atomic</i></td></tr>
<tr><th id="103">103</th><td><em>static</em> <b>inline</b>  <em>int</em> <dfn class="tu decl def" id="__cxxabiv1::decrementHandlerCount" title='__cxxabiv1::decrementHandlerCount' data-type='int __cxxabiv1::decrementHandlerCount(__cxxabiv1::__cxa_exception * exception)' data-ref="__cxxabiv1::decrementHandlerCount">decrementHandlerCount</dfn>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a> *<dfn class="local col0 decl" id="10exception" title='exception' data-type='__cxxabiv1::__cxa_exception *' data-ref="10exception">exception</dfn>) {</td></tr>
<tr><th id="104">104</th><td>    <b>return</b> --<a class="local col0 ref" href="#10exception" title='exception' data-ref="10exception">exception</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a>;</td></tr>
<tr><th id="105">105</th><td>}</td></tr>
<tr><th id="106">106</th><td></td></tr>
<tr><th id="107">107</th><td><u>#include <a href="fallback_malloc.ipp.html">"fallback_malloc.ipp"</a></u></td></tr>
<tr><th id="108">108</th><td></td></tr>
<tr><th id="109">109</th><td><i  data-doc="__cxxabiv1::do_malloc">//  Allocate some memory from _somewhere_</i></td></tr>
<tr><th id="110">110</th><td><em>static</em> <em>void</em> *<dfn class="tu decl def" id="__cxxabiv1::do_malloc" title='__cxxabiv1::do_malloc' data-type='void * __cxxabiv1::do_malloc(size_t size)' data-ref="__cxxabiv1::do_malloc">do_malloc</dfn>(<span class='typedef' title='size_t' data-type='unsigned long' data-ref="size_t">size_t</span> <dfn class="local col0 decl" id="30size" title='size' data-type='size_t' data-ref="30size">size</dfn>) {</td></tr>
<tr><th id="111">111</th><td>    <em>void</em> *<dfn class="local col1 decl" id="31ptr" title='ptr' data-type='void *' data-ref="31ptr">ptr</dfn> = <span class="namespace">std::</span><a class="ref" href="../../../bionic/libc/include/malloc.h.html#malloc" title='malloc' data-ref="malloc">malloc</a>(<a class="local col0 ref" href="#30size" title='size' data-ref="30size">size</a>);</td></tr>
<tr><th id="112">112</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col1 ref" href="#31ptr" title='ptr' data-ref="31ptr">ptr</a>) <i>// if malloc fails, fall back to emergency stash</i></td></tr>
<tr><th id="113">113</th><td>        <a class="local col1 ref" href="#31ptr" title='ptr' data-ref="31ptr">ptr</a> = <a class="tu ref" href="fallback_malloc.ipp.html#__cxxabiv1::(anonymousnamespace)::fallback_malloc" title='__cxxabiv1::(anonymous namespace)::fallback_malloc' data-use='c' data-ref="__cxxabiv1::(anonymousnamespace)::fallback_malloc">fallback_malloc</a>(<a class="local col0 ref" href="#30size" title='size' data-ref="30size">size</a>);</td></tr>
<tr><th id="114">114</th><td>    <b>return</b> <a class="local col1 ref" href="#31ptr" title='ptr' data-ref="31ptr">ptr</a>;</td></tr>
<tr><th id="115">115</th><td>}</td></tr>
<tr><th id="116">116</th><td></td></tr>
<tr><th id="117">117</th><td><em>static</em> <em>void</em> <dfn class="tu decl def" id="__cxxabiv1::do_free" title='__cxxabiv1::do_free' data-type='void __cxxabiv1::do_free(void * ptr)' data-ref="__cxxabiv1::do_free">do_free</dfn>(<em>void</em> *<dfn class="local col2 decl" id="32ptr" title='ptr' data-type='void *' data-ref="32ptr">ptr</dfn>) {</td></tr>
<tr><th id="118">118</th><td>    <a class="tu ref" href="fallback_malloc.ipp.html#__cxxabiv1::(anonymousnamespace)::is_fallback_ptr" title='__cxxabiv1::(anonymous namespace)::is_fallback_ptr' data-use='c' data-ref="__cxxabiv1::(anonymousnamespace)::is_fallback_ptr">is_fallback_ptr</a>(<a class="local col2 ref" href="#32ptr" title='ptr' data-ref="32ptr">ptr</a>) ? <a class="tu ref" href="fallback_malloc.ipp.html#__cxxabiv1::(anonymousnamespace)::fallback_free" title='__cxxabiv1::(anonymous namespace)::fallback_free' data-use='c' data-ref="__cxxabiv1::(anonymousnamespace)::fallback_free">fallback_free</a>(<a class="local col2 ref" href="#32ptr" title='ptr' data-ref="32ptr">ptr</a>) : <span class="namespace">std::</span><a class="ref" href="../../../bionic/libc/include/malloc.h.html#free" title='free' data-ref="free">free</a>(<a class="local col2 ref" href="#32ptr" title='ptr' data-ref="32ptr">ptr</a>);</td></tr>
<tr><th id="119">119</th><td>}</td></tr>
<tr><th id="120">120</th><td></td></tr>
<tr><th id="121">121</th><td><i  data-doc="__cxxabiv1::exception_cleanup_func">/*</i></td></tr>
<tr><th id="122">122</th><td><i  data-doc="__cxxabiv1::exception_cleanup_func">    If reason isn't _URC_FOREIGN_EXCEPTION_CAUGHT, then the terminateHandler</i></td></tr>
<tr><th id="123">123</th><td><i  data-doc="__cxxabiv1::exception_cleanup_func">    stored in exc is called.  Otherwise the exceptionDestructor stored in </i></td></tr>
<tr><th id="124">124</th><td><i  data-doc="__cxxabiv1::exception_cleanup_func">    exc is called, and then the memory for the exception is deallocated.</i></td></tr>
<tr><th id="125">125</th><td><i  data-doc="__cxxabiv1::exception_cleanup_func"></i></td></tr>
<tr><th id="126">126</th><td><i  data-doc="__cxxabiv1::exception_cleanup_func">    This is never called for a __cxa_dependent_exception.</i></td></tr>
<tr><th id="127">127</th><td><i  data-doc="__cxxabiv1::exception_cleanup_func">*/</i></td></tr>
<tr><th id="128">128</th><td><em>static</em></td></tr>
<tr><th id="129">129</th><td><em>void</em></td></tr>
<tr><th id="130">130</th><td><dfn class="tu decl def" id="__cxxabiv1::exception_cleanup_func" title='__cxxabiv1::exception_cleanup_func' data-type='void __cxxabiv1::exception_cleanup_func(_Unwind_Reason_Code reason, _Unwind_Exception * unwind_exception)' data-ref="__cxxabiv1::exception_cleanup_func">exception_cleanup_func</dfn>(<span class='typedef' title='_Unwind_Reason_Code' data-type='enum _Unwind_Reason_Code' data-ref="_Unwind_Reason_Code">_Unwind_Reason_Code</span> <dfn class="local col3 decl" id="33reason" title='reason' data-type='_Unwind_Reason_Code' data-ref="33reason">reason</dfn>, <span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col4 decl" id="34unwind_exception" title='unwind_exception' data-type='_Unwind_Exception *' data-ref="34unwind_exception">unwind_exception</dfn>)</td></tr>
<tr><th id="131">131</th><td>{</td></tr>
<tr><th id="132">132</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col5 decl" id="35exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="35exception_header">exception_header</dfn> = <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_exception_unwind_exception" title='__cxxabiv1::cxa_exception_from_exception_unwind_exception' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_exception_unwind_exception">cxa_exception_from_exception_unwind_exception</a>(<a class="local col4 ref" href="#34unwind_exception" title='unwind_exception' data-ref="34unwind_exception">unwind_exception</a>);</td></tr>
<tr><th id="133">133</th><td>    <b>if</b> (<span class='enum' title='_URC_FOREIGN_EXCEPTION_CAUGHT' data-ref="_URC_FOREIGN_EXCEPTION_CAUGHT">_URC_FOREIGN_EXCEPTION_CAUGHT</span> != <a class="local col3 ref" href="#33reason" title='reason' data-ref="33reason">reason</a>)</td></tr>
<tr><th id="134">134</th><td>        <span class="namespace">std::</span><a class="ref" href="cxa_handlers.hpp.html#_ZSt11__terminatePFvvE" title='std::__terminate' data-ref="_ZSt11__terminatePFvvE">__terminate</a>(<a class="local col5 ref" href="#35exception_header" title='exception_header' data-ref="35exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::terminateHandler" title='__cxxabiv1::__cxa_exception::terminateHandler' data-ref="__cxxabiv1::__cxa_exception::terminateHandler">terminateHandler</a>);</td></tr>
<tr><th id="135">135</th><td>    <i>// Just in case there exists a dependent exception that is pointing to this,</i></td></tr>
<tr><th id="136">136</th><td><i>    //    check the reference count and only destroy this if that count goes to zero.</i></td></tr>
<tr><th id="137">137</th><td>    <a class="ref" href="#__cxa_decrement_exception_refcount" title='__cxa_decrement_exception_refcount' data-ref="__cxa_decrement_exception_refcount">__cxa_decrement_exception_refcount</a>(<a class="local col4 ref" href="#34unwind_exception" title='unwind_exception' data-ref="34unwind_exception">unwind_exception</a> + <var>1</var>);</td></tr>
<tr><th id="138">138</th><td>}</td></tr>
<tr><th id="139">139</th><td></td></tr>
<tr><th id="140">140</th><td><em>static</em> <a class="macro" href="../include/cxxabi.h.html#24" title="__attribute__((noreturn))" data-ref="_M/LIBCXXABI_NORETURN">LIBCXXABI_NORETURN</a> <em>void</em> <dfn class="tu decl def" id="__cxxabiv1::failed_throw" title='__cxxabiv1::failed_throw' data-type='void __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception * exception_header)' data-ref="__cxxabiv1::failed_throw">failed_throw</dfn>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col6 decl" id="36exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="36exception_header">exception_header</dfn>) {</td></tr>
<tr><th id="141">141</th><td><i>//  Section 2.5.3 says:</i></td></tr>
<tr><th id="142">142</th><td><i>//      * For purposes of this ABI, several things are considered exception handlers:</i></td></tr>
<tr><th id="143">143</th><td><i>//      ** A terminate() call due to a throw.</i></td></tr>
<tr><th id="144">144</th><td><i>//  and</i></td></tr>
<tr><th id="145">145</th><td><i>//      * Upon entry, Following initialization of the catch parameter, </i></td></tr>
<tr><th id="146">146</th><td><i>//          a handler must call:</i></td></tr>
<tr><th id="147">147</th><td><i>//      * void *__cxa_begin_catch(void *exceptionObject );</i></td></tr>
<tr><th id="148">148</th><td>    (<em>void</em>) <a class="ref" href="#__cxa_begin_catch" title='__cxa_begin_catch' data-ref="__cxa_begin_catch">__cxa_begin_catch</a>(&amp;<a class="local col6 ref" href="#36exception_header" title='exception_header' data-ref="36exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="149">149</th><td>    <span class="namespace">std::</span><a class="ref" href="cxa_handlers.hpp.html#_ZSt11__terminatePFvvE" title='std::__terminate' data-ref="_ZSt11__terminatePFvvE">__terminate</a>(<a class="local col6 ref" href="#36exception_header" title='exception_header' data-ref="36exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::terminateHandler" title='__cxxabiv1::__cxa_exception::terminateHandler' data-ref="__cxxabiv1::__cxa_exception::terminateHandler">terminateHandler</a>);</td></tr>
<tr><th id="150">150</th><td>}</td></tr>
<tr><th id="151">151</th><td></td></tr>
<tr><th id="152">152</th><td><b>extern</b> <q>"C"</q> {</td></tr>
<tr><th id="153">153</th><td></td></tr>
<tr><th id="154">154</th><td><i>//  Allocate a __cxa_exception object, and zero-fill it.</i></td></tr>
<tr><th id="155">155</th><td><i>//  Reserve "thrown_size" bytes on the end for the user's exception</i></td></tr>
<tr><th id="156">156</th><td><i>//  object. Zero-fill the object. If memory can't be allocated, call</i></td></tr>
<tr><th id="157">157</th><td><i>//  std::terminate. Return a pointer to the memory to be used for the</i></td></tr>
<tr><th id="158">158</th><td><i>//  user's exception object.</i></td></tr>
<tr><th id="159">159</th><td><em>void</em> * <dfn class="decl def" id="__cxa_allocate_exception" title='__cxa_allocate_exception' data-ref="__cxa_allocate_exception">__cxa_allocate_exception</dfn> (<span class='typedef' title='size_t' data-type='unsigned long' data-ref="size_t">size_t</span> <dfn class="local col7 decl" id="37thrown_size" title='thrown_size' data-type='size_t' data-ref="37thrown_size">thrown_size</dfn>) <b>throw</b>() {</td></tr>
<tr><th id="160">160</th><td>    <span class='typedef' title='size_t' data-type='unsigned long' data-ref="size_t">size_t</span> <dfn class="local col8 decl" id="38actual_size" title='actual_size' data-type='size_t' data-ref="38actual_size">actual_size</dfn> = <a class="tu ref" href="#__cxxabiv1::cxa_exception_size_from_exception_thrown_size" title='__cxxabiv1::cxa_exception_size_from_exception_thrown_size' data-use='c' data-ref="__cxxabiv1::cxa_exception_size_from_exception_thrown_size">cxa_exception_size_from_exception_thrown_size</a>(<a class="local col7 ref" href="#37thrown_size" title='thrown_size' data-ref="37thrown_size">thrown_size</a>);</td></tr>
<tr><th id="161">161</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col9 decl" id="39exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="39exception_header">exception_header</dfn> = <b>static_cast</b>&lt;<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>*&gt;(<a class="tu ref" href="#__cxxabiv1::do_malloc" title='__cxxabiv1::do_malloc' data-use='c' data-ref="__cxxabiv1::do_malloc">do_malloc</a>(<a class="local col8 ref" href="#38actual_size" title='actual_size' data-ref="38actual_size">actual_size</a>));</td></tr>
<tr><th id="162">162</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col9 ref" href="#39exception_header" title='exception_header' data-ref="39exception_header">exception_header</a>)</td></tr>
<tr><th id="163">163</th><td>        <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt9terminatev" title='std::terminate' data-ref="_ZSt9terminatev">terminate</a>();</td></tr>
<tr><th id="164">164</th><td>    <span class="namespace">std::</span><a class="ref" href="../../../bionic/libc/include/string.h.html#memset" title='memset' data-ref="memset">memset</a>(<a class="local col9 ref" href="#39exception_header" title='exception_header' data-ref="39exception_header">exception_header</a>, <var>0</var>, <a class="local col8 ref" href="#38actual_size" title='actual_size' data-ref="38actual_size">actual_size</a>);</td></tr>
<tr><th id="165">165</th><td>    <b>return</b> <a class="tu ref" href="#__cxxabiv1::thrown_object_from_cxa_exception" title='__cxxabiv1::thrown_object_from_cxa_exception' data-use='c' data-ref="__cxxabiv1::thrown_object_from_cxa_exception">thrown_object_from_cxa_exception</a>(<a class="local col9 ref" href="#39exception_header" title='exception_header' data-ref="39exception_header">exception_header</a>);</td></tr>
<tr><th id="166">166</th><td>}</td></tr>
<tr><th id="167">167</th><td></td></tr>
<tr><th id="168">168</th><td></td></tr>
<tr><th id="169">169</th><td><i>//  Free a __cxa_exception object allocated with __cxa_allocate_exception.</i></td></tr>
<tr><th id="170">170</th><td><em>void</em> <dfn class="decl def" id="__cxa_free_exception" title='__cxa_free_exception' data-ref="__cxa_free_exception">__cxa_free_exception</dfn> (<em>void</em> * <dfn class="local col0 decl" id="40thrown_object" title='thrown_object' data-type='void *' data-ref="40thrown_object">thrown_object</dfn>) <b>throw</b>() {</td></tr>
<tr><th id="171">171</th><td>    <a class="tu ref" href="#__cxxabiv1::do_free" title='__cxxabiv1::do_free' data-use='c' data-ref="__cxxabiv1::do_free">do_free</a>(<a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col0 ref" href="#40thrown_object" title='thrown_object' data-ref="40thrown_object">thrown_object</a>));</td></tr>
<tr><th id="172">172</th><td>}</td></tr>
<tr><th id="173">173</th><td></td></tr>
<tr><th id="174">174</th><td></td></tr>
<tr><th id="175">175</th><td><i>//  This function shall allocate a __cxa_dependent_exception and</i></td></tr>
<tr><th id="176">176</th><td><i>//  return a pointer to it. (Really to the object, not past its' end).</i></td></tr>
<tr><th id="177">177</th><td><i>//  Otherwise, it will work like __cxa_allocate_exception.</i></td></tr>
<tr><th id="178">178</th><td><em>void</em> * <dfn class="decl def" id="__cxa_allocate_dependent_exception" title='__cxa_allocate_dependent_exception' data-ref="__cxa_allocate_dependent_exception">__cxa_allocate_dependent_exception</dfn> () {</td></tr>
<tr><th id="179">179</th><td>    <span class='typedef' title='size_t' data-type='unsigned long' data-ref="size_t">size_t</span> <dfn class="local col1 decl" id="41actual_size" title='actual_size' data-type='size_t' data-ref="41actual_size">actual_size</dfn> = <b>sizeof</b>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>);</td></tr>
<tr><th id="180">180</th><td>    <em>void</em> *<dfn class="local col2 decl" id="42ptr" title='ptr' data-type='void *' data-ref="42ptr">ptr</dfn> = <a class="tu ref" href="#__cxxabiv1::do_malloc" title='__cxxabiv1::do_malloc' data-use='c' data-ref="__cxxabiv1::do_malloc">do_malloc</a>(<a class="local col1 ref" href="#41actual_size" title='actual_size' data-ref="41actual_size">actual_size</a>);</td></tr>
<tr><th id="181">181</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col2 ref" href="#42ptr" title='ptr' data-ref="42ptr">ptr</a>)</td></tr>
<tr><th id="182">182</th><td>        <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt9terminatev" title='std::terminate' data-ref="_ZSt9terminatev">terminate</a>();</td></tr>
<tr><th id="183">183</th><td>    <span class="namespace">std::</span><a class="ref" href="../../../bionic/libc/include/string.h.html#memset" title='memset' data-ref="memset">memset</a>(<a class="local col2 ref" href="#42ptr" title='ptr' data-ref="42ptr">ptr</a>, <var>0</var>, <a class="local col1 ref" href="#41actual_size" title='actual_size' data-ref="41actual_size">actual_size</a>);</td></tr>
<tr><th id="184">184</th><td>    <b>return</b> <a class="local col2 ref" href="#42ptr" title='ptr' data-ref="42ptr">ptr</a>;</td></tr>
<tr><th id="185">185</th><td>}</td></tr>
<tr><th id="186">186</th><td></td></tr>
<tr><th id="187">187</th><td></td></tr>
<tr><th id="188">188</th><td><i>//  This function shall free a dependent_exception.</i></td></tr>
<tr><th id="189">189</th><td><i>//  It does not affect the reference count of the primary exception.</i></td></tr>
<tr><th id="190">190</th><td><em>void</em> <dfn class="decl def" id="__cxa_free_dependent_exception" title='__cxa_free_dependent_exception' data-ref="__cxa_free_dependent_exception">__cxa_free_dependent_exception</dfn> (<em>void</em> * <dfn class="local col3 decl" id="43dependent_exception" title='dependent_exception' data-type='void *' data-ref="43dependent_exception">dependent_exception</dfn>) {</td></tr>
<tr><th id="191">191</th><td>    <a class="tu ref" href="#__cxxabiv1::do_free" title='__cxxabiv1::do_free' data-use='c' data-ref="__cxxabiv1::do_free">do_free</a>(<a class="local col3 ref" href="#43dependent_exception" title='dependent_exception' data-ref="43dependent_exception">dependent_exception</a>);</td></tr>
<tr><th id="192">192</th><td>}</td></tr>
<tr><th id="193">193</th><td></td></tr>
<tr><th id="194">194</th><td></td></tr>
<tr><th id="195">195</th><td><i>// 2.4.3 Throwing the Exception Object</i></td></tr>
<tr><th id="196">196</th><td><i>/*</i></td></tr>
<tr><th id="197">197</th><td><i>After constructing the exception object with the throw argument value,</i></td></tr>
<tr><th id="198">198</th><td><i>the generated code calls the __cxa_throw runtime library routine. This</i></td></tr>
<tr><th id="199">199</th><td><i>routine never returns.</i></td></tr>
<tr><th id="200">200</th><td><i></i></td></tr>
<tr><th id="201">201</th><td><i>The __cxa_throw routine will do the following:</i></td></tr>
<tr><th id="202">202</th><td><i></i></td></tr>
<tr><th id="203">203</th><td><i>* Obtain the __cxa_exception header from the thrown exception object address,</i></td></tr>
<tr><th id="204">204</th><td><i>which can be computed as follows: </i></td></tr>
<tr><th id="205">205</th><td><i> __cxa_exception *header = ((__cxa_exception *) thrown_exception - 1); </i></td></tr>
<tr><th id="206">206</th><td><i>* Save the current unexpected_handler and terminate_handler in the __cxa_exception header.</i></td></tr>
<tr><th id="207">207</th><td><i>* Save the tinfo and dest arguments in the __cxa_exception header. </i></td></tr>
<tr><th id="208">208</th><td><i>* Set the exception_class field in the unwind header. This is a 64-bit value</i></td></tr>
<tr><th id="209">209</th><td><i>representing the ASCII string "XXXXC++\0", where "XXXX" is a</i></td></tr>
<tr><th id="210">210</th><td><i>vendor-dependent string. That is, for implementations conforming to this</i></td></tr>
<tr><th id="211">211</th><td><i>ABI, the low-order 4 bytes of this 64-bit value will be "C++\0".</i></td></tr>
<tr><th id="212">212</th><td><i>* Increment the uncaught_exception flag. </i></td></tr>
<tr><th id="213">213</th><td><i>* Call _Unwind_RaiseException in the system unwind library, Its argument is the</i></td></tr>
<tr><th id="214">214</th><td><i>pointer to the thrown exception, which __cxa_throw itself received as an argument.</i></td></tr>
<tr><th id="215">215</th><td><i>__Unwind_RaiseException begins the process of stack unwinding, described</i></td></tr>
<tr><th id="216">216</th><td><i>in Section 2.5. In special cases, such as an inability to find a</i></td></tr>
<tr><th id="217">217</th><td><i>handler, _Unwind_RaiseException may return. In that case, __cxa_throw</i></td></tr>
<tr><th id="218">218</th><td><i>will call terminate, assuming that there was no handler for the</i></td></tr>
<tr><th id="219">219</th><td><i>exception.</i></td></tr>
<tr><th id="220">220</th><td><i>*/</i></td></tr>
<tr><th id="221">221</th><td><a class="macro" href="../include/cxxabi.h.html#24" title="__attribute__((noreturn))" data-ref="_M/LIBCXXABI_NORETURN">LIBCXXABI_NORETURN</a></td></tr>
<tr><th id="222">222</th><td><em>void</em> </td></tr>
<tr><th id="223">223</th><td><dfn class="decl def" id="__cxa_throw" title='__cxa_throw' data-ref="__cxa_throw">__cxa_throw</dfn>(<em>void</em>* <dfn class="local col4 decl" id="44thrown_object" title='thrown_object' data-type='void *' data-ref="44thrown_object">thrown_object</dfn>, <span class="namespace">std::</span><a class="type" href="../include/cxxabi.h.html#std::type_info" title='std::type_info' data-ref="std::type_info">type_info</a>* <dfn class="local col5 decl" id="45tinfo" title='tinfo' data-type='std::type_info *' data-ref="45tinfo">tinfo</dfn>, <em>void</em> (*<dfn class="local col6 decl" id="46dest" title='dest' data-type='void (*)(void *)' data-ref="46dest">dest</dfn>)(<em>void</em>*))</td></tr>
<tr><th id="224">224</th><td>{</td></tr>
<tr><th id="225">225</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals" title='__cxxabiv1::__cxa_eh_globals' data-ref="__cxxabiv1::__cxa_eh_globals">__cxa_eh_globals</a> *<dfn class="local col7 decl" id="47globals" title='globals' data-type='__cxxabiv1::__cxa_eh_globals *' data-ref="47globals">globals</dfn> = <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals" title='__cxa_get_globals' data-ref="__cxa_get_globals">__cxa_get_globals</a>();</td></tr>
<tr><th id="226">226</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col8 decl" id="48exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="48exception_header">exception_header</dfn> = <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col4 ref" href="#44thrown_object" title='thrown_object' data-ref="44thrown_object">thrown_object</a>);</td></tr>
<tr><th id="227">227</th><td></td></tr>
<tr><th id="228">228</th><td>    <a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unexpectedHandler" title='__cxxabiv1::__cxa_exception::unexpectedHandler' data-ref="__cxxabiv1::__cxa_exception::unexpectedHandler">unexpectedHandler</a> = <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt14get_unexpectedv" title='std::get_unexpected' data-ref="_ZSt14get_unexpectedv">get_unexpected</a>();</td></tr>
<tr><th id="229">229</th><td>    <a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::terminateHandler" title='__cxxabiv1::__cxa_exception::terminateHandler' data-ref="__cxxabiv1::__cxa_exception::terminateHandler">terminateHandler</a>  = <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt13get_terminatev" title='std::get_terminate' data-ref="_ZSt13get_terminatev">get_terminate</a>();</td></tr>
<tr><th id="230">230</th><td>    <a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::exceptionType" title='__cxxabiv1::__cxa_exception::exceptionType' data-ref="__cxxabiv1::__cxa_exception::exceptionType">exceptionType</a> = <a class="local col5 ref" href="#45tinfo" title='tinfo' data-ref="45tinfo">tinfo</a>;</td></tr>
<tr><th id="231">231</th><td>    <a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::exceptionDestructor" title='__cxxabiv1::__cxa_exception::exceptionDestructor' data-ref="__cxxabiv1::__cxa_exception::exceptionDestructor">exceptionDestructor</a> = <a class="local col6 ref" href="#46dest" title='dest' data-ref="46dest">dest</a>;</td></tr>
<tr><th id="232">232</th><td>    <a class="tu ref" href="#__cxxabiv1::setExceptionClass" title='__cxxabiv1::setExceptionClass' data-use='c' data-ref="__cxxabiv1::setExceptionClass">setExceptionClass</a>(&amp;<a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="233">233</th><td>    <a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::referenceCount" title='__cxxabiv1::__cxa_exception::referenceCount' data-ref="__cxxabiv1::__cxa_exception::referenceCount">referenceCount</a> = <var>1</var>;  <i>// This is a newly allocated exception, no need for thread safety.</i></td></tr>
<tr><th id="234">234</th><td>    <a class="local col7 ref" href="#47globals" title='globals' data-ref="47globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::uncaughtExceptions" title='__cxxabiv1::__cxa_eh_globals::uncaughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::uncaughtExceptions">uncaughtExceptions</a> += <var>1</var>;   <i>// Not atomically, since globals are thread-local</i></td></tr>
<tr><th id="235">235</th><td></td></tr>
<tr><th id="236">236</th><td>    <a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>.<span class='ref' title='_Unwind_Exception::exception_cleanup' data-ref="_Unwind_Exception::exception_cleanup">exception_cleanup</span> = <a class="tu ref" href="#__cxxabiv1::exception_cleanup_func" title='__cxxabiv1::exception_cleanup_func' data-use='r' data-ref="__cxxabiv1::exception_cleanup_func">exception_cleanup_func</a>;</td></tr>
<tr><th id="237">237</th><td><u>#<span data-ppcond="237">ifdef</span> <span class="macro" data-ref="_M/__USING_SJLJ_EXCEPTIONS__">__USING_SJLJ_EXCEPTIONS__</span></u></td></tr>
<tr><th id="238">238</th><td>    _Unwind_SjLj_RaiseException(&amp;exception_header-&gt;unwindHeader);</td></tr>
<tr><th id="239">239</th><td><u>#<span data-ppcond="237">else</span></u></td></tr>
<tr><th id="240">240</th><td>    <span class='ref' title='_Unwind_RaiseException' data-ref="_Unwind_RaiseException">_Unwind_RaiseException</span>(&amp;<a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="241">241</th><td><u>#<span data-ppcond="237">endif</span></u></td></tr>
<tr><th id="242">242</th><td>    <i>//  This only happens when there is no handler, or some unexpected unwinding</i></td></tr>
<tr><th id="243">243</th><td><i>    //     error happens.</i></td></tr>
<tr><th id="244">244</th><td>    <a class="tu ref" href="#__cxxabiv1::failed_throw" title='__cxxabiv1::failed_throw' data-use='c' data-ref="__cxxabiv1::failed_throw">failed_throw</a>(<a class="local col8 ref" href="#48exception_header" title='exception_header' data-ref="48exception_header">exception_header</a>);</td></tr>
<tr><th id="245">245</th><td>}</td></tr>
<tr><th id="246">246</th><td></td></tr>
<tr><th id="247">247</th><td></td></tr>
<tr><th id="248">248</th><td><i>// 2.5.3 Exception Handlers</i></td></tr>
<tr><th id="249">249</th><td><i>/*</i></td></tr>
<tr><th id="250">250</th><td><i>The adjusted pointer is computed by the personality routine during phase 1</i></td></tr>
<tr><th id="251">251</th><td><i>  and saved in the exception header (either __cxa_exception or</i></td></tr>
<tr><th id="252">252</th><td><i>  __cxa_dependent_exception).</i></td></tr>
<tr><th id="253">253</th><td><i></i></td></tr>
<tr><th id="254">254</th><td><i>  Requires:  exception is native</i></td></tr>
<tr><th id="255">255</th><td><i>*/</i></td></tr>
<tr><th id="256">256</th><td><em>void</em>*</td></tr>
<tr><th id="257">257</th><td><dfn class="decl def" id="__cxa_get_exception_ptr" title='__cxa_get_exception_ptr' data-ref="__cxa_get_exception_ptr">__cxa_get_exception_ptr</dfn>(<em>void</em>* <dfn class="local col9 decl" id="49unwind_exception" title='unwind_exception' data-type='void *' data-ref="49unwind_exception">unwind_exception</dfn>) <b>throw</b>()</td></tr>
<tr><th id="258">258</th><td>{</td></tr>
<tr><th id="259">259</th><td><u>#<span data-ppcond="259">if</span> <a class="macro" href="../include/__cxxabi_config.h.html#17" title="0" data-ref="_M/LIBCXXABI_ARM_EHABI">LIBCXXABI_ARM_EHABI</a></u></td></tr>
<tr><th id="260">260</th><td>    <b>return</b> <b>reinterpret_cast</b>&lt;<em>void</em>*&gt;(</td></tr>
<tr><th id="261">261</th><td>        <b>static_cast</b>&lt;_Unwind_Control_Block*&gt;(unwind_exception)-&gt;barrier_cache.bitpattern[<var>0</var>]);</td></tr>
<tr><th id="262">262</th><td><u>#<span data-ppcond="259">else</span></u></td></tr>
<tr><th id="263">263</th><td>    <b>return</b> <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_exception_unwind_exception" title='__cxxabiv1::cxa_exception_from_exception_unwind_exception' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_exception_unwind_exception">cxa_exception_from_exception_unwind_exception</a>(</td></tr>
<tr><th id="264">264</th><td>        <b>static_cast</b>&lt;<span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>*&gt;(<a class="local col9 ref" href="#49unwind_exception" title='unwind_exception' data-ref="49unwind_exception">unwind_exception</a>))-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::adjustedPtr" title='__cxxabiv1::__cxa_exception::adjustedPtr' data-ref="__cxxabiv1::__cxa_exception::adjustedPtr">adjustedPtr</a>;</td></tr>
<tr><th id="265">265</th><td><u>#<span data-ppcond="259">endif</span></u></td></tr>
<tr><th id="266">266</th><td>}</td></tr>
<tr><th id="267">267</th><td></td></tr>
<tr><th id="268">268</th><td><u>#<span data-ppcond="268">if</span> <a class="macro" href="../include/__cxxabi_config.h.html#17" title="0" data-ref="_M/LIBCXXABI_ARM_EHABI">LIBCXXABI_ARM_EHABI</a></u></td></tr>
<tr><th id="269">269</th><td><i>/*</i></td></tr>
<tr><th id="270">270</th><td><i>The routine to be called before the cleanup.  This will save __cxa_exception in</i></td></tr>
<tr><th id="271">271</th><td><i>__cxa_eh_globals, so that __cxa_end_cleanup() can recover later.</i></td></tr>
<tr><th id="272">272</th><td><i>*/</i></td></tr>
<tr><th id="273">273</th><td><em>bool</em></td></tr>
<tr><th id="274">274</th><td>__cxa_begin_cleanup(<em>void</em>* unwind_arg) <b>throw</b> ()</td></tr>
<tr><th id="275">275</th><td>{</td></tr>
<tr><th id="276">276</th><td>    _Unwind_Exception* unwind_exception = <b>static_cast</b>&lt;_Unwind_Exception*&gt;(unwind_arg);</td></tr>
<tr><th id="277">277</th><td>    __cxa_eh_globals* globals = __cxa_get_globals();</td></tr>
<tr><th id="278">278</th><td>    __cxa_exception* exception_header =</td></tr>
<tr><th id="279">279</th><td>        cxa_exception_from_exception_unwind_exception(unwind_exception);</td></tr>
<tr><th id="280">280</th><td></td></tr>
<tr><th id="281">281</th><td>    <b>if</b> (isOurExceptionClass(unwind_exception))</td></tr>
<tr><th id="282">282</th><td>    {</td></tr>
<tr><th id="283">283</th><td>        <b>if</b> (<var>0</var> == exception_header-&gt;propagationCount)</td></tr>
<tr><th id="284">284</th><td>        {</td></tr>
<tr><th id="285">285</th><td>            exception_header-&gt;nextPropagatingException = globals-&gt;propagatingExceptions;</td></tr>
<tr><th id="286">286</th><td>            globals-&gt;propagatingExceptions = exception_header;</td></tr>
<tr><th id="287">287</th><td>        }</td></tr>
<tr><th id="288">288</th><td>        ++exception_header-&gt;propagationCount;</td></tr>
<tr><th id="289">289</th><td>    }</td></tr>
<tr><th id="290">290</th><td>    <b>else</b></td></tr>
<tr><th id="291">291</th><td>    {</td></tr>
<tr><th id="292">292</th><td>        <i>// If the propagatingExceptions stack is not empty, since we can't</i></td></tr>
<tr><th id="293">293</th><td><i>        // chain the foreign exception, terminate it.</i></td></tr>
<tr><th id="294">294</th><td>        <b>if</b> (NULL != globals-&gt;propagatingExceptions)</td></tr>
<tr><th id="295">295</th><td>            std::terminate();</td></tr>
<tr><th id="296">296</th><td>        globals-&gt;propagatingExceptions = exception_header;</td></tr>
<tr><th id="297">297</th><td>    }</td></tr>
<tr><th id="298">298</th><td>    <b>return</b> <b>true</b>;</td></tr>
<tr><th id="299">299</th><td>}</td></tr>
<tr><th id="300">300</th><td></td></tr>
<tr><th id="301">301</th><td><i>/*</i></td></tr>
<tr><th id="302">302</th><td><i>The routine to be called after the cleanup has been performed.  It will get the</i></td></tr>
<tr><th id="303">303</th><td><i>propagating __cxa_exception from __cxa_eh_globals, and continue the stack</i></td></tr>
<tr><th id="304">304</th><td><i>unwinding with _Unwind_Resume.</i></td></tr>
<tr><th id="305">305</th><td><i></i></td></tr>
<tr><th id="306">306</th><td><i>According to ARM EHABI 8.4.1, __cxa_end_cleanup() should not clobber any</i></td></tr>
<tr><th id="307">307</th><td><i>register, thus we have to write this function in assembly so that we can save</i></td></tr>
<tr><th id="308">308</th><td><i>{r1, r2, r3}.  We don't have to save r0 because it is the return value and the</i></td></tr>
<tr><th id="309">309</th><td><i>first argument to _Unwind_Resume().  In addition, we are saving r4 in order to</i></td></tr>
<tr><th id="310">310</th><td><i>align the stack to 16 bytes, even though it is a callee-save register.</i></td></tr>
<tr><th id="311">311</th><td><i>*/</i></td></tr>
<tr><th id="312">312</th><td><b>__attribute__</b>((used)) <em>static</em> _Unwind_Exception *</td></tr>
<tr><th id="313">313</th><td>__cxa_end_cleanup_impl()</td></tr>
<tr><th id="314">314</th><td>{</td></tr>
<tr><th id="315">315</th><td>    __cxa_eh_globals* globals = __cxa_get_globals();</td></tr>
<tr><th id="316">316</th><td>    __cxa_exception* exception_header = globals-&gt;propagatingExceptions;</td></tr>
<tr><th id="317">317</th><td>    <b>if</b> (NULL == exception_header)</td></tr>
<tr><th id="318">318</th><td>    {</td></tr>
<tr><th id="319">319</th><td>        <i>// It seems that __cxa_begin_cleanup() is not called properly.</i></td></tr>
<tr><th id="320">320</th><td><i>        // We have no choice but terminate the program now.</i></td></tr>
<tr><th id="321">321</th><td>        std::terminate();</td></tr>
<tr><th id="322">322</th><td>    }</td></tr>
<tr><th id="323">323</th><td></td></tr>
<tr><th id="324">324</th><td>    <b>if</b> (isOurExceptionClass(&amp;exception_header-&gt;unwindHeader))</td></tr>
<tr><th id="325">325</th><td>    {</td></tr>
<tr><th id="326">326</th><td>        --exception_header-&gt;propagationCount;</td></tr>
<tr><th id="327">327</th><td>        <b>if</b> (<var>0</var> == exception_header-&gt;propagationCount)</td></tr>
<tr><th id="328">328</th><td>        {</td></tr>
<tr><th id="329">329</th><td>            globals-&gt;propagatingExceptions = exception_header-&gt;nextPropagatingException;</td></tr>
<tr><th id="330">330</th><td>            exception_header-&gt;nextPropagatingException = NULL;</td></tr>
<tr><th id="331">331</th><td>        }</td></tr>
<tr><th id="332">332</th><td>    }</td></tr>
<tr><th id="333">333</th><td>    <b>else</b></td></tr>
<tr><th id="334">334</th><td>    {</td></tr>
<tr><th id="335">335</th><td>        globals-&gt;propagatingExceptions = NULL;</td></tr>
<tr><th id="336">336</th><td>    }</td></tr>
<tr><th id="337">337</th><td>    <b>return</b> &amp;exception_header-&gt;unwindHeader;</td></tr>
<tr><th id="338">338</th><td>}</td></tr>
<tr><th id="339">339</th><td></td></tr>
<tr><th id="340">340</th><td><b>asm</b> (</td></tr>
<tr><th id="341">341</th><td>    <q>"	.pushsection	.text.__cxa_end_cleanup,\"ax\",%progbits\n"</q></td></tr>
<tr><th id="342">342</th><td>    <q>"	.globl	__cxa_end_cleanup\n"</q></td></tr>
<tr><th id="343">343</th><td>    <q>"	.type	__cxa_end_cleanup,%function\n"</q></td></tr>
<tr><th id="344">344</th><td>    <q>"__cxa_end_cleanup:\n"</q></td></tr>
<tr><th id="345">345</th><td>    <q>"	push	{r1, r2, r3, r4}\n"</q></td></tr>
<tr><th id="346">346</th><td>    <q>"	bl	__cxa_end_cleanup_impl\n"</q></td></tr>
<tr><th id="347">347</th><td>    <q>"	pop	{r1, r2, r3, r4}\n"</q></td></tr>
<tr><th id="348">348</th><td>    <q>"	bl	_Unwind_Resume\n"</q></td></tr>
<tr><th id="349">349</th><td>    <q>"	bl	abort\n"</q></td></tr>
<tr><th id="350">350</th><td>    <q>"	.popsection"</q></td></tr>
<tr><th id="351">351</th><td>);</td></tr>
<tr><th id="352">352</th><td><u>#<span data-ppcond="268">endif</span>  // LIBCXXABI_ARM_EHABI</u></td></tr>
<tr><th id="353">353</th><td>    </td></tr>
<tr><th id="354">354</th><td><i>/*</i></td></tr>
<tr><th id="355">355</th><td><i>This routine can catch foreign or native exceptions.  If native, the exception</i></td></tr>
<tr><th id="356">356</th><td><i>can be a primary or dependent variety.  This routine may remain blissfully</i></td></tr>
<tr><th id="357">357</th><td><i>ignorant of whether the native exception is primary or dependent.</i></td></tr>
<tr><th id="358">358</th><td><i></i></td></tr>
<tr><th id="359">359</th><td><i>If the exception is native:</i></td></tr>
<tr><th id="360">360</th><td><i>* Increment's the exception's handler count.</i></td></tr>
<tr><th id="361">361</th><td><i>* Push the exception on the stack of currently-caught exceptions if it is not </i></td></tr>
<tr><th id="362">362</th><td><i>  already there (from a rethrow).</i></td></tr>
<tr><th id="363">363</th><td><i>* Decrements the uncaught_exception count.</i></td></tr>
<tr><th id="364">364</th><td><i>* Returns the adjusted pointer to the exception object, which is stored in</i></td></tr>
<tr><th id="365">365</th><td><i>  the __cxa_exception by the personality routine.</i></td></tr>
<tr><th id="366">366</th><td><i></i></td></tr>
<tr><th id="367">367</th><td><i>If the exception is foreign, this means it did not originate from one of throw</i></td></tr>
<tr><th id="368">368</th><td><i>routines.  The foreign exception does not necessarily have a __cxa_exception</i></td></tr>
<tr><th id="369">369</th><td><i>header.  However we can catch it here with a catch (...), or with a call</i></td></tr>
<tr><th id="370">370</th><td><i>to terminate or unexpected during unwinding.</i></td></tr>
<tr><th id="371">371</th><td><i>* Do not try to increment the exception's handler count, we don't know where</i></td></tr>
<tr><th id="372">372</th><td><i>  it is.</i></td></tr>
<tr><th id="373">373</th><td><i>* Push the exception on the stack of currently-caught exceptions only if the</i></td></tr>
<tr><th id="374">374</th><td><i>  stack is empty.  The foreign exception has no way to link to the current</i></td></tr>
<tr><th id="375">375</th><td><i>  top of stack.  If the stack is not empty, call terminate.  Even with an</i></td></tr>
<tr><th id="376">376</th><td><i>  empty stack, this is hacked in by pushing a pointer to an imaginary</i></td></tr>
<tr><th id="377">377</th><td><i>  __cxa_exception block in front of the foreign exception.  It would be better</i></td></tr>
<tr><th id="378">378</th><td><i>  if the __cxa_eh_globals structure had a stack of _Unwind_Exception, but it</i></td></tr>
<tr><th id="379">379</th><td><i>  doesn't.  It has a stack of __cxa_exception (which has a next* in it).</i></td></tr>
<tr><th id="380">380</th><td><i>* Do not decrement the uncaught_exception count because we didn't increment it</i></td></tr>
<tr><th id="381">381</th><td><i>  in __cxa_throw (or one of our rethrow functions).</i></td></tr>
<tr><th id="382">382</th><td><i>* If we haven't terminated, assume the exception object is just past the </i></td></tr>
<tr><th id="383">383</th><td><i>  _Unwind_Exception and return a pointer to that.</i></td></tr>
<tr><th id="384">384</th><td><i>*/</i></td></tr>
<tr><th id="385">385</th><td><em>void</em>*</td></tr>
<tr><th id="386">386</th><td><dfn class="decl def" id="__cxa_begin_catch" title='__cxa_begin_catch' data-ref="__cxa_begin_catch">__cxa_begin_catch</dfn>(<em>void</em>* <dfn class="local col0 decl" id="50unwind_arg" title='unwind_arg' data-type='void *' data-ref="50unwind_arg">unwind_arg</dfn>) <b>throw</b>()</td></tr>
<tr><th id="387">387</th><td>{</td></tr>
<tr><th id="388">388</th><td>    <span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col1 decl" id="51unwind_exception" title='unwind_exception' data-type='_Unwind_Exception *' data-ref="51unwind_exception">unwind_exception</dfn> = <b>static_cast</b>&lt;<span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>*&gt;(<a class="local col0 ref" href="#50unwind_arg" title='unwind_arg' data-ref="50unwind_arg">unwind_arg</a>);</td></tr>
<tr><th id="389">389</th><td>    <em>bool</em> <dfn class="local col2 decl" id="52native_exception" title='native_exception' data-type='bool' data-ref="52native_exception">native_exception</dfn> = <a class="tu ref" href="#__cxxabiv1::isOurExceptionClass" title='__cxxabiv1::isOurExceptionClass' data-use='c' data-ref="__cxxabiv1::isOurExceptionClass">isOurExceptionClass</a>(<a class="local col1 ref" href="#51unwind_exception" title='unwind_exception' data-ref="51unwind_exception">unwind_exception</a>);</td></tr>
<tr><th id="390">390</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals" title='__cxxabiv1::__cxa_eh_globals' data-ref="__cxxabiv1::__cxa_eh_globals">__cxa_eh_globals</a>* <dfn class="local col3 decl" id="53globals" title='globals' data-type='__cxxabiv1::__cxa_eh_globals *' data-ref="53globals">globals</dfn> = <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals" title='__cxa_get_globals' data-ref="__cxa_get_globals">__cxa_get_globals</a>();</td></tr>
<tr><th id="391">391</th><td>    <i>// exception_header is a hackish offset from a foreign exception, but it</i></td></tr>
<tr><th id="392">392</th><td><i>    //   works as long as we're careful not to try to access any __cxa_exception</i></td></tr>
<tr><th id="393">393</th><td><i>    //   parts.</i></td></tr>
<tr><th id="394">394</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col4 decl" id="54exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="54exception_header">exception_header</dfn> =</td></tr>
<tr><th id="395">395</th><td>            <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_exception_unwind_exception" title='__cxxabiv1::cxa_exception_from_exception_unwind_exception' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_exception_unwind_exception">cxa_exception_from_exception_unwind_exception</a></td></tr>
<tr><th id="396">396</th><td>            (</td></tr>
<tr><th id="397">397</th><td>                <b>static_cast</b>&lt;<span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>*&gt;(<a class="local col1 ref" href="#51unwind_exception" title='unwind_exception' data-ref="51unwind_exception">unwind_exception</a>)</td></tr>
<tr><th id="398">398</th><td>            );</td></tr>
<tr><th id="399">399</th><td>    <b>if</b> (<a class="local col2 ref" href="#52native_exception" title='native_exception' data-ref="52native_exception">native_exception</a>)</td></tr>
<tr><th id="400">400</th><td>    {</td></tr>
<tr><th id="401">401</th><td>        <i>// Increment the handler count, removing the flag about being rethrown</i></td></tr>
<tr><th id="402">402</th><td>        <a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a> = <a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a> &lt; <var>0</var> ?</td></tr>
<tr><th id="403">403</th><td>            -<a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a> + <var>1</var> : <a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a> + <var>1</var>;</td></tr>
<tr><th id="404">404</th><td>        <i>//  place the exception on the top of the stack if it's not already</i></td></tr>
<tr><th id="405">405</th><td><i>        //    there by a previous rethrow</i></td></tr>
<tr><th id="406">406</th><td>        <b>if</b> (<a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a> != <a class="local col3 ref" href="#53globals" title='globals' data-ref="53globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a>)</td></tr>
<tr><th id="407">407</th><td>        {</td></tr>
<tr><th id="408">408</th><td>            <a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::nextException" title='__cxxabiv1::__cxa_exception::nextException' data-ref="__cxxabiv1::__cxa_exception::nextException">nextException</a> = <a class="local col3 ref" href="#53globals" title='globals' data-ref="53globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a>;</td></tr>
<tr><th id="409">409</th><td>            <a class="local col3 ref" href="#53globals" title='globals' data-ref="53globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a> = <a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>;</td></tr>
<tr><th id="410">410</th><td>        }</td></tr>
<tr><th id="411">411</th><td>        <a class="local col3 ref" href="#53globals" title='globals' data-ref="53globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::uncaughtExceptions" title='__cxxabiv1::__cxa_eh_globals::uncaughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::uncaughtExceptions">uncaughtExceptions</a> -= <var>1</var>;   <i>// Not atomically, since globals are thread-local</i></td></tr>
<tr><th id="412">412</th><td><u>#<span data-ppcond="412">if</span> <a class="macro" href="../include/__cxxabi_config.h.html#17" title="0" data-ref="_M/LIBCXXABI_ARM_EHABI">LIBCXXABI_ARM_EHABI</a></u></td></tr>
<tr><th id="413">413</th><td>        <b>return</b> <b>reinterpret_cast</b>&lt;<em>void</em>*&gt;(exception_header-&gt;unwindHeader.barrier_cache.bitpattern[<var>0</var>]);</td></tr>
<tr><th id="414">414</th><td><u>#<span data-ppcond="412">else</span></u></td></tr>
<tr><th id="415">415</th><td>        <b>return</b> <a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::adjustedPtr" title='__cxxabiv1::__cxa_exception::adjustedPtr' data-ref="__cxxabiv1::__cxa_exception::adjustedPtr">adjustedPtr</a>;</td></tr>
<tr><th id="416">416</th><td><u>#<span data-ppcond="412">endif</span></u></td></tr>
<tr><th id="417">417</th><td>    }</td></tr>
<tr><th id="418">418</th><td>    <i>// Else this is a foreign exception</i></td></tr>
<tr><th id="419">419</th><td><i>    // If the caughtExceptions stack is not empty, terminate</i></td></tr>
<tr><th id="420">420</th><td>    <b>if</b> (<a class="local col3 ref" href="#53globals" title='globals' data-ref="53globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a> != <var>0</var>)</td></tr>
<tr><th id="421">421</th><td>        <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt9terminatev" title='std::terminate' data-ref="_ZSt9terminatev">terminate</a>();</td></tr>
<tr><th id="422">422</th><td>    <i>// Push the foreign exception on to the stack</i></td></tr>
<tr><th id="423">423</th><td>    <a class="local col3 ref" href="#53globals" title='globals' data-ref="53globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a> = <a class="local col4 ref" href="#54exception_header" title='exception_header' data-ref="54exception_header">exception_header</a>;</td></tr>
<tr><th id="424">424</th><td>    <b>return</b> <a class="local col1 ref" href="#51unwind_exception" title='unwind_exception' data-ref="51unwind_exception">unwind_exception</a> + <var>1</var>;</td></tr>
<tr><th id="425">425</th><td>}</td></tr>
<tr><th id="426">426</th><td></td></tr>
<tr><th id="427">427</th><td></td></tr>
<tr><th id="428">428</th><td><i>/*</i></td></tr>
<tr><th id="429">429</th><td><i>Upon exit for any reason, a handler must call:</i></td></tr>
<tr><th id="430">430</th><td><i>    void __cxa_end_catch ();</i></td></tr>
<tr><th id="431">431</th><td><i></i></td></tr>
<tr><th id="432">432</th><td><i>This routine can be called for either a native or foreign exception.</i></td></tr>
<tr><th id="433">433</th><td><i>For a native exception:</i></td></tr>
<tr><th id="434">434</th><td><i>* Locates the most recently caught exception and decrements its handler count.</i></td></tr>
<tr><th id="435">435</th><td><i>* Removes the exception from the caught exception stack, if the handler count goes to zero.</i></td></tr>
<tr><th id="436">436</th><td><i>* If the handler count goes down to zero, and the exception was not re-thrown</i></td></tr>
<tr><th id="437">437</th><td><i>  by throw, it locates the primary exception (which may be the same as the one</i></td></tr>
<tr><th id="438">438</th><td><i>  it's handling) and decrements its reference count. If that reference count</i></td></tr>
<tr><th id="439">439</th><td><i>  goes to zero, the function destroys the exception. In any case, if the current</i></td></tr>
<tr><th id="440">440</th><td><i>  exception is a dependent exception, it destroys that.</i></td></tr>
<tr><th id="441">441</th><td><i></i></td></tr>
<tr><th id="442">442</th><td><i>For a foreign exception:</i></td></tr>
<tr><th id="443">443</th><td><i>* If it has been rethrown, there is nothing to do.</i></td></tr>
<tr><th id="444">444</th><td><i>* Otherwise delete the exception and pop the catch stack to empty.</i></td></tr>
<tr><th id="445">445</th><td><i>*/</i></td></tr>
<tr><th id="446">446</th><td><em>void</em> <dfn class="decl def" id="__cxa_end_catch" title='__cxa_end_catch' data-ref="__cxa_end_catch">__cxa_end_catch</dfn>()</td></tr>
<tr><th id="447">447</th><td>{</td></tr>
<tr><th id="448">448</th><td>  <b>static_assert</b>(<b>sizeof</b>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>) == <b>sizeof</b>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>),</td></tr>
<tr><th id="449">449</th><td>                <q>"sizeof(__cxa_exception) must be equal to "</q></td></tr>
<tr><th id="450">450</th><td>                <q>"sizeof(__cxa_dependent_exception)"</q>);</td></tr>
<tr><th id="451">451</th><td>    <b>static_assert</b>(<span class="macro" title="__builtin_offsetof(__cxa_exception, referenceCount)" data-ref="_M/offsetof">offsetof</span>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>, referenceCount) ==</td></tr>
<tr><th id="452">452</th><td>                      <span class="macro" title="__builtin_offsetof(__cxa_dependent_exception, primaryException)" data-ref="_M/offsetof">offsetof</span>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>, primaryException),</td></tr>
<tr><th id="453">453</th><td>                  <q>"the layout of __cxa_exception must match the layout of "</q></td></tr>
<tr><th id="454">454</th><td>                  <q>"__cxa_dependent_exception"</q>);</td></tr>
<tr><th id="455">455</th><td>    <b>static_assert</b>(<span class="macro" title="__builtin_offsetof(__cxa_exception, handlerCount)" data-ref="_M/offsetof">offsetof</span>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>, handlerCount) ==</td></tr>
<tr><th id="456">456</th><td>                      <span class="macro" title="__builtin_offsetof(__cxa_dependent_exception, handlerCount)" data-ref="_M/offsetof">offsetof</span>(<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>, handlerCount),</td></tr>
<tr><th id="457">457</th><td>                  <q>"the layout of __cxa_exception must match the layout of "</q></td></tr>
<tr><th id="458">458</th><td>                  <q>"__cxa_dependent_exception"</q>);</td></tr>
<tr><th id="459">459</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals" title='__cxxabiv1::__cxa_eh_globals' data-ref="__cxxabiv1::__cxa_eh_globals">__cxa_eh_globals</a>* <dfn class="local col5 decl" id="55globals" title='globals' data-type='__cxxabiv1::__cxa_eh_globals *' data-ref="55globals">globals</dfn> = <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals_fast" title='__cxa_get_globals_fast' data-ref="__cxa_get_globals_fast">__cxa_get_globals_fast</a>(); <i>// __cxa_get_globals called in __cxa_begin_catch</i></td></tr>
<tr><th id="460">460</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col6 decl" id="56exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="56exception_header">exception_header</dfn> = <a class="local col5 ref" href="#55globals" title='globals' data-ref="55globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a>;</td></tr>
<tr><th id="461">461</th><td>    <i>// If we've rethrown a foreign exception, then globals-&gt;caughtExceptions</i></td></tr>
<tr><th id="462">462</th><td><i>    //    will have been made an empty stack by __cxa_rethrow() and there is</i></td></tr>
<tr><th id="463">463</th><td><i>    //    nothing more to be done.  Do nothing!</i></td></tr>
<tr><th id="464">464</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> != <a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>)</td></tr>
<tr><th id="465">465</th><td>    {</td></tr>
<tr><th id="466">466</th><td>        <em>bool</em> <dfn class="local col7 decl" id="57native_exception" title='native_exception' data-type='bool' data-ref="57native_exception">native_exception</dfn> = <a class="tu ref" href="#__cxxabiv1::isOurExceptionClass" title='__cxxabiv1::isOurExceptionClass' data-use='c' data-ref="__cxxabiv1::isOurExceptionClass">isOurExceptionClass</a>(&amp;<a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="467">467</th><td>        <b>if</b> (<a class="local col7 ref" href="#57native_exception" title='native_exception' data-ref="57native_exception">native_exception</a>)</td></tr>
<tr><th id="468">468</th><td>        {</td></tr>
<tr><th id="469">469</th><td>            <i>// This is a native exception</i></td></tr>
<tr><th id="470">470</th><td>            <b>if</b> (<a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a> &lt; <var>0</var>)</td></tr>
<tr><th id="471">471</th><td>            {</td></tr>
<tr><th id="472">472</th><td>                <i>//  The exception has been rethrown by __cxa_rethrow, so don't delete it</i></td></tr>
<tr><th id="473">473</th><td>                <b>if</b> (<var>0</var> == <a class="tu ref" href="#__cxxabiv1::incrementHandlerCount" title='__cxxabiv1::incrementHandlerCount' data-use='c' data-ref="__cxxabiv1::incrementHandlerCount">incrementHandlerCount</a>(<a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>))</td></tr>
<tr><th id="474">474</th><td>                {</td></tr>
<tr><th id="475">475</th><td>                    <i>//  Remove from the chain of uncaught exceptions</i></td></tr>
<tr><th id="476">476</th><td>                    <a class="local col5 ref" href="#55globals" title='globals' data-ref="55globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a> = <a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::nextException" title='__cxxabiv1::__cxa_exception::nextException' data-ref="__cxxabiv1::__cxa_exception::nextException">nextException</a>;</td></tr>
<tr><th id="477">477</th><td>                    <i>// but don't destroy</i></td></tr>
<tr><th id="478">478</th><td>                }</td></tr>
<tr><th id="479">479</th><td>                <i>// Keep handlerCount negative in case there are nested catch's</i></td></tr>
<tr><th id="480">480</th><td><i>                //   that need to be told that this exception is rethrown.  Don't</i></td></tr>
<tr><th id="481">481</th><td><i>                //   erase this rethrow flag until the exception is recaught.</i></td></tr>
<tr><th id="482">482</th><td>            }</td></tr>
<tr><th id="483">483</th><td>            <b>else</b></td></tr>
<tr><th id="484">484</th><td>            {</td></tr>
<tr><th id="485">485</th><td>                <i>// The native exception has not been rethrown</i></td></tr>
<tr><th id="486">486</th><td>                <b>if</b> (<var>0</var> == <a class="tu ref" href="#__cxxabiv1::decrementHandlerCount" title='__cxxabiv1::decrementHandlerCount' data-use='c' data-ref="__cxxabiv1::decrementHandlerCount">decrementHandlerCount</a>(<a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>))</td></tr>
<tr><th id="487">487</th><td>                {</td></tr>
<tr><th id="488">488</th><td>                    <i>//  Remove from the chain of uncaught exceptions</i></td></tr>
<tr><th id="489">489</th><td>                    <a class="local col5 ref" href="#55globals" title='globals' data-ref="55globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a> = <a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::nextException" title='__cxxabiv1::__cxa_exception::nextException' data-ref="__cxxabiv1::__cxa_exception::nextException">nextException</a>;</td></tr>
<tr><th id="490">490</th><td>                    <i>// Destroy this exception, being careful to distinguish</i></td></tr>
<tr><th id="491">491</th><td><i>                    //    between dependent and primary exceptions</i></td></tr>
<tr><th id="492">492</th><td>                    <b>if</b> (<a class="tu ref" href="#__cxxabiv1::isDependentException" title='__cxxabiv1::isDependentException' data-use='c' data-ref="__cxxabiv1::isDependentException">isDependentException</a>(&amp;<a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>))</td></tr>
<tr><th id="493">493</th><td>                    {</td></tr>
<tr><th id="494">494</th><td>                        <i>// Reset exception_header to primaryException and deallocate the dependent exception</i></td></tr>
<tr><th id="495">495</th><td>                        <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>* <dfn class="local col8 decl" id="58dep_exception_header" title='dep_exception_header' data-type='__cxxabiv1::__cxa_dependent_exception *' data-ref="58dep_exception_header">dep_exception_header</dfn> =</td></tr>
<tr><th id="496">496</th><td>                            <b>reinterpret_cast</b>&lt;<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>*&gt;(<a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>);</td></tr>
<tr><th id="497">497</th><td>                        <a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a> =</td></tr>
<tr><th id="498">498</th><td>                            <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col8 ref" href="#58dep_exception_header" title='dep_exception_header' data-ref="58dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::primaryException" title='__cxxabiv1::__cxa_dependent_exception::primaryException' data-ref="__cxxabiv1::__cxa_dependent_exception::primaryException">primaryException</a>);</td></tr>
<tr><th id="499">499</th><td>                        <a class="ref" href="#__cxa_free_dependent_exception" title='__cxa_free_dependent_exception' data-ref="__cxa_free_dependent_exception">__cxa_free_dependent_exception</a>(<a class="local col8 ref" href="#58dep_exception_header" title='dep_exception_header' data-ref="58dep_exception_header">dep_exception_header</a>);</td></tr>
<tr><th id="500">500</th><td>                    }</td></tr>
<tr><th id="501">501</th><td>                    <i>// Destroy the primary exception only if its referenceCount goes to 0</i></td></tr>
<tr><th id="502">502</th><td><i>                    //    (this decrement must be atomic)</i></td></tr>
<tr><th id="503">503</th><td>                    <a class="ref" href="#__cxa_decrement_exception_refcount" title='__cxa_decrement_exception_refcount' data-ref="__cxa_decrement_exception_refcount">__cxa_decrement_exception_refcount</a>(<a class="tu ref" href="#__cxxabiv1::thrown_object_from_cxa_exception" title='__cxxabiv1::thrown_object_from_cxa_exception' data-use='c' data-ref="__cxxabiv1::thrown_object_from_cxa_exception">thrown_object_from_cxa_exception</a>(<a class="local col6 ref" href="#56exception_header" title='exception_header' data-ref="56exception_header">exception_header</a>));</td></tr>
<tr><th id="504">504</th><td>                }</td></tr>
<tr><th id="505">505</th><td>            }</td></tr>
<tr><th id="506">506</th><td>        }       </td></tr>
<tr><th id="507">507</th><td>        <b>else</b></td></tr>
<tr><th id="508">508</th><td>        {</td></tr>
<tr><th id="509">509</th><td>            <i>// The foreign exception has not been rethrown.  Pop the stack</i></td></tr>
<tr><th id="510">510</th><td><i>            //    and delete it.  If there are nested catch's and they try</i></td></tr>
<tr><th id="511">511</th><td><i>            //    to touch a foreign exception in any way, that is undefined</i></td></tr>
<tr><th id="512">512</th><td><i>            //     behavior.  They likely can't since the only way to catch</i></td></tr>
<tr><th id="513">513</th><td><i>            //     a foreign exception is with catch (...)!</i></td></tr>
<tr><th id="514">514</th><td>            <span class='ref' title='_Unwind_DeleteException' data-ref="_Unwind_DeleteException">_Unwind_DeleteException</span>(&amp;<a class="local col5 ref" href="#55globals" title='globals' data-ref="55globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="515">515</th><td>            <a class="local col5 ref" href="#55globals" title='globals' data-ref="55globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a> = <var>0</var>;</td></tr>
<tr><th id="516">516</th><td>        }</td></tr>
<tr><th id="517">517</th><td>    }</td></tr>
<tr><th id="518">518</th><td>}</td></tr>
<tr><th id="519">519</th><td></td></tr>
<tr><th id="520">520</th><td><i>// Note:  exception_header may be masquerading as a __cxa_dependent_exception</i></td></tr>
<tr><th id="521">521</th><td><i>//        and that's ok.  exceptionType is there too.</i></td></tr>
<tr><th id="522">522</th><td><i>//        However watch out for foreign exceptions.  Return null for them.</i></td></tr>
<tr><th id="523">523</th><td><span class="namespace">std::</span><a class="type" href="../include/cxxabi.h.html#std::type_info" title='std::type_info' data-ref="std::type_info">type_info</a> * <dfn class="decl def" id="__cxa_current_exception_type" title='__cxa_current_exception_type' data-ref="__cxa_current_exception_type">__cxa_current_exception_type</dfn>() {</td></tr>
<tr><th id="524">524</th><td><i>//  get the current exception</i></td></tr>
<tr><th id="525">525</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals" title='__cxxabiv1::__cxa_eh_globals' data-ref="__cxxabiv1::__cxa_eh_globals">__cxa_eh_globals</a> *<dfn class="local col9 decl" id="59globals" title='globals' data-type='__cxxabiv1::__cxa_eh_globals *' data-ref="59globals">globals</dfn> = <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals_fast" title='__cxa_get_globals_fast' data-ref="__cxa_get_globals_fast">__cxa_get_globals_fast</a>();</td></tr>
<tr><th id="526">526</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col9 ref" href="#59globals" title='globals' data-ref="59globals">globals</a>)</td></tr>
<tr><th id="527">527</th><td>        <b>return</b> <span class="macro" title="__null" data-ref="_M/NULL">NULL</span>;     <i>//  If there have never been any exceptions, there are none now.</i></td></tr>
<tr><th id="528">528</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a> *<dfn class="local col0 decl" id="60exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="60exception_header">exception_header</dfn> = <a class="local col9 ref" href="#59globals" title='globals' data-ref="59globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a>;</td></tr>
<tr><th id="529">529</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col0 ref" href="#60exception_header" title='exception_header' data-ref="60exception_header">exception_header</a>)</td></tr>
<tr><th id="530">530</th><td>        <b>return</b> <span class="macro" title="__null" data-ref="_M/NULL">NULL</span>;        <i>//  No current exception</i></td></tr>
<tr><th id="531">531</th><td>    <b>if</b> (!<a class="tu ref" href="#__cxxabiv1::isOurExceptionClass" title='__cxxabiv1::isOurExceptionClass' data-use='c' data-ref="__cxxabiv1::isOurExceptionClass">isOurExceptionClass</a>(&amp;<a class="local col0 ref" href="#60exception_header" title='exception_header' data-ref="60exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>))</td></tr>
<tr><th id="532">532</th><td>        <b>return</b> <span class="macro" title="__null" data-ref="_M/NULL">NULL</span>;</td></tr>
<tr><th id="533">533</th><td>    <b>return</b> <a class="local col0 ref" href="#60exception_header" title='exception_header' data-ref="60exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::exceptionType" title='__cxxabiv1::__cxa_exception::exceptionType' data-ref="__cxxabiv1::__cxa_exception::exceptionType">exceptionType</a>;</td></tr>
<tr><th id="534">534</th><td>}</td></tr>
<tr><th id="535">535</th><td></td></tr>
<tr><th id="536">536</th><td><i>// 2.5.4 Rethrowing Exceptions</i></td></tr>
<tr><th id="537">537</th><td><i>/*  This routine can rethrow native or foreign exceptions.</i></td></tr>
<tr><th id="538">538</th><td><i>If the exception is native:</i></td></tr>
<tr><th id="539">539</th><td><i>* marks the exception object on top of the caughtExceptions stack </i></td></tr>
<tr><th id="540">540</th><td><i>  (in an implementation-defined way) as being rethrown. </i></td></tr>
<tr><th id="541">541</th><td><i>* If the caughtExceptions stack is empty, it calls terminate() </i></td></tr>
<tr><th id="542">542</th><td><i>  (see [C++FDIS] [except.throw], 15.1.8). </i></td></tr>
<tr><th id="543">543</th><td><i>* It then calls _Unwind_RaiseException which should not return</i></td></tr>
<tr><th id="544">544</th><td><i>   (terminate if it does).</i></td></tr>
<tr><th id="545">545</th><td><i>  Note:  exception_header may be masquerading as a __cxa_dependent_exception</i></td></tr>
<tr><th id="546">546</th><td><i>         and that's ok.</i></td></tr>
<tr><th id="547">547</th><td><i>*/</i></td></tr>
<tr><th id="548">548</th><td><a class="macro" href="../include/cxxabi.h.html#24" title="__attribute__((noreturn))" data-ref="_M/LIBCXXABI_NORETURN">LIBCXXABI_NORETURN</a></td></tr>
<tr><th id="549">549</th><td><em>void</em></td></tr>
<tr><th id="550">550</th><td><dfn class="decl def" id="__cxa_rethrow" title='__cxa_rethrow' data-ref="__cxa_rethrow">__cxa_rethrow</dfn>()</td></tr>
<tr><th id="551">551</th><td>{</td></tr>
<tr><th id="552">552</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals" title='__cxxabiv1::__cxa_eh_globals' data-ref="__cxxabiv1::__cxa_eh_globals">__cxa_eh_globals</a>* <dfn class="local col1 decl" id="61globals" title='globals' data-type='__cxxabiv1::__cxa_eh_globals *' data-ref="61globals">globals</dfn> = <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals" title='__cxa_get_globals' data-ref="__cxa_get_globals">__cxa_get_globals</a>();</td></tr>
<tr><th id="553">553</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col2 decl" id="62exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="62exception_header">exception_header</dfn> = <a class="local col1 ref" href="#61globals" title='globals' data-ref="61globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a>;</td></tr>
<tr><th id="554">554</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col2 ref" href="#62exception_header" title='exception_header' data-ref="62exception_header">exception_header</a>)</td></tr>
<tr><th id="555">555</th><td>        <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt9terminatev" title='std::terminate' data-ref="_ZSt9terminatev">terminate</a>();      <i>// throw; called outside of a exception handler</i></td></tr>
<tr><th id="556">556</th><td>    <em>bool</em> <dfn class="local col3 decl" id="63native_exception" title='native_exception' data-type='bool' data-ref="63native_exception">native_exception</dfn> = <a class="tu ref" href="#__cxxabiv1::isOurExceptionClass" title='__cxxabiv1::isOurExceptionClass' data-use='c' data-ref="__cxxabiv1::isOurExceptionClass">isOurExceptionClass</a>(&amp;<a class="local col2 ref" href="#62exception_header" title='exception_header' data-ref="62exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="557">557</th><td>    <b>if</b> (<a class="local col3 ref" href="#63native_exception" title='native_exception' data-ref="63native_exception">native_exception</a>)</td></tr>
<tr><th id="558">558</th><td>    {</td></tr>
<tr><th id="559">559</th><td>        <i>//  Mark the exception as being rethrown (reverse the effects of __cxa_begin_catch)</i></td></tr>
<tr><th id="560">560</th><td>        <a class="local col2 ref" href="#62exception_header" title='exception_header' data-ref="62exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a> = -<a class="local col2 ref" href="#62exception_header" title='exception_header' data-ref="62exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::handlerCount" title='__cxxabiv1::__cxa_exception::handlerCount' data-ref="__cxxabiv1::__cxa_exception::handlerCount">handlerCount</a>;</td></tr>
<tr><th id="561">561</th><td>        <a class="local col1 ref" href="#61globals" title='globals' data-ref="61globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::uncaughtExceptions" title='__cxxabiv1::__cxa_eh_globals::uncaughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::uncaughtExceptions">uncaughtExceptions</a> += <var>1</var>;</td></tr>
<tr><th id="562">562</th><td>        <i>//  __cxa_end_catch will remove this exception from the caughtExceptions stack if necessary</i></td></tr>
<tr><th id="563">563</th><td>    }</td></tr>
<tr><th id="564">564</th><td>    <b>else</b>  <i>// this is a foreign exception</i></td></tr>
<tr><th id="565">565</th><td>    {</td></tr>
<tr><th id="566">566</th><td>        <i>// The only way to communicate to __cxa_end_catch that we've rethrown</i></td></tr>
<tr><th id="567">567</th><td><i>        //   a foreign exception, so don't delete us, is to pop the stack here</i></td></tr>
<tr><th id="568">568</th><td><i>        //   which must be empty afterwards.  Then __cxa_end_catch will do</i></td></tr>
<tr><th id="569">569</th><td><i>        //   nothing</i></td></tr>
<tr><th id="570">570</th><td>        <a class="local col1 ref" href="#61globals" title='globals' data-ref="61globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a> = <var>0</var>;</td></tr>
<tr><th id="571">571</th><td>    }</td></tr>
<tr><th id="572">572</th><td><u>#<span data-ppcond="572">ifdef</span> <span class="macro" data-ref="_M/__USING_SJLJ_EXCEPTIONS__">__USING_SJLJ_EXCEPTIONS__</span></u></td></tr>
<tr><th id="573">573</th><td>    _Unwind_SjLj_RaiseException(&amp;exception_header-&gt;unwindHeader);</td></tr>
<tr><th id="574">574</th><td><u>#<span data-ppcond="572">else</span></u></td></tr>
<tr><th id="575">575</th><td>    <span class='ref' title='_Unwind_RaiseException' data-ref="_Unwind_RaiseException">_Unwind_RaiseException</span>(&amp;<a class="local col2 ref" href="#62exception_header" title='exception_header' data-ref="62exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="576">576</th><td><u>#<span data-ppcond="572">endif</span></u></td></tr>
<tr><th id="577">577</th><td></td></tr>
<tr><th id="578">578</th><td>    <i>//  If we get here, some kind of unwinding error has occurred.</i></td></tr>
<tr><th id="579">579</th><td><i>    //  There is some weird code generation bug happening with </i></td></tr>
<tr><th id="580">580</th><td><i>    //     Apple clang version 4.0 (tags/Apple/clang-418.0.2) (based on LLVM 3.1svn)</i></td></tr>
<tr><th id="581">581</th><td><i>    //     If we call failed_throw here.  Turns up with -O2 or higher, and -Os.</i></td></tr>
<tr><th id="582">582</th><td>    <a class="ref" href="#__cxa_begin_catch" title='__cxa_begin_catch' data-ref="__cxa_begin_catch">__cxa_begin_catch</a>(&amp;<a class="local col2 ref" href="#62exception_header" title='exception_header' data-ref="62exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="583">583</th><td>    <b>if</b> (<a class="local col3 ref" href="#63native_exception" title='native_exception' data-ref="63native_exception">native_exception</a>)</td></tr>
<tr><th id="584">584</th><td>        <span class="namespace">std::</span><a class="ref" href="cxa_handlers.hpp.html#_ZSt11__terminatePFvvE" title='std::__terminate' data-ref="_ZSt11__terminatePFvvE">__terminate</a>(<a class="local col2 ref" href="#62exception_header" title='exception_header' data-ref="62exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::terminateHandler" title='__cxxabiv1::__cxa_exception::terminateHandler' data-ref="__cxxabiv1::__cxa_exception::terminateHandler">terminateHandler</a>);</td></tr>
<tr><th id="585">585</th><td>    <i>// Foreign exception: can't get exception_header-&gt;terminateHandler</i></td></tr>
<tr><th id="586">586</th><td>    <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt9terminatev" title='std::terminate' data-ref="_ZSt9terminatev">terminate</a>();</td></tr>
<tr><th id="587">587</th><td>}</td></tr>
<tr><th id="588">588</th><td></td></tr>
<tr><th id="589">589</th><td><i>/*</i></td></tr>
<tr><th id="590">590</th><td><i>    If thrown_object is not null, atomically increment the referenceCount field</i></td></tr>
<tr><th id="591">591</th><td><i>    of the __cxa_exception header associated with the thrown object referred to</i></td></tr>
<tr><th id="592">592</th><td><i>    by thrown_object.</i></td></tr>
<tr><th id="593">593</th><td><i></i></td></tr>
<tr><th id="594">594</th><td><i>    Requires:  If thrown_object is not NULL, it is a native exception.</i></td></tr>
<tr><th id="595">595</th><td><i>*/</i></td></tr>
<tr><th id="596">596</th><td><em>void</em></td></tr>
<tr><th id="597">597</th><td><dfn class="decl def" id="__cxa_increment_exception_refcount" title='__cxa_increment_exception_refcount' data-ref="__cxa_increment_exception_refcount">__cxa_increment_exception_refcount</dfn>(<em>void</em>* <dfn class="local col4 decl" id="64thrown_object" title='thrown_object' data-type='void *' data-ref="64thrown_object">thrown_object</dfn>) <b>throw</b>()</td></tr>
<tr><th id="598">598</th><td>{</td></tr>
<tr><th id="599">599</th><td>    <b>if</b> (<a class="local col4 ref" href="#64thrown_object" title='thrown_object' data-ref="64thrown_object">thrown_object</a> != <span class="macro" title="__null" data-ref="_M/NULL">NULL</span> )</td></tr>
<tr><th id="600">600</th><td>    {</td></tr>
<tr><th id="601">601</th><td>        <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col5 decl" id="65exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="65exception_header">exception_header</dfn> = <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col4 ref" href="#64thrown_object" title='thrown_object' data-ref="64thrown_object">thrown_object</a>);</td></tr>
<tr><th id="602">602</th><td>        <a class="ref" href="#602" title='__sync_add_and_fetch_8' data-ref="__sync_add_and_fetch_8">__sync_add_and_fetch</a>(&amp;<a class="local col5 ref" href="#65exception_header" title='exception_header' data-ref="65exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::referenceCount" title='__cxxabiv1::__cxa_exception::referenceCount' data-ref="__cxxabiv1::__cxa_exception::referenceCount">referenceCount</a>, <var>1</var>);</td></tr>
<tr><th id="603">603</th><td>    }</td></tr>
<tr><th id="604">604</th><td>}</td></tr>
<tr><th id="605">605</th><td></td></tr>
<tr><th id="606">606</th><td><i>/*</i></td></tr>
<tr><th id="607">607</th><td><i>    If thrown_object is not null, atomically decrement the referenceCount field</i></td></tr>
<tr><th id="608">608</th><td><i>    of the __cxa_exception header associated with the thrown object referred to</i></td></tr>
<tr><th id="609">609</th><td><i>    by thrown_object.  If the referenceCount drops to zero, destroy and</i></td></tr>
<tr><th id="610">610</th><td><i>    deallocate the exception.</i></td></tr>
<tr><th id="611">611</th><td><i></i></td></tr>
<tr><th id="612">612</th><td><i>    Requires:  If thrown_object is not NULL, it is a native exception.</i></td></tr>
<tr><th id="613">613</th><td><i>*/</i></td></tr>
<tr><th id="614">614</th><td><em>void</em></td></tr>
<tr><th id="615">615</th><td><dfn class="decl def" id="__cxa_decrement_exception_refcount" title='__cxa_decrement_exception_refcount' data-ref="__cxa_decrement_exception_refcount">__cxa_decrement_exception_refcount</dfn>(<em>void</em>* <dfn class="local col6 decl" id="66thrown_object" title='thrown_object' data-type='void *' data-ref="66thrown_object">thrown_object</dfn>) <b>throw</b>()</td></tr>
<tr><th id="616">616</th><td>{</td></tr>
<tr><th id="617">617</th><td>    <b>if</b> (<a class="local col6 ref" href="#66thrown_object" title='thrown_object' data-ref="66thrown_object">thrown_object</a> != <span class="macro" title="__null" data-ref="_M/NULL">NULL</span> )</td></tr>
<tr><th id="618">618</th><td>    {</td></tr>
<tr><th id="619">619</th><td>        <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col7 decl" id="67exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="67exception_header">exception_header</dfn> = <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col6 ref" href="#66thrown_object" title='thrown_object' data-ref="66thrown_object">thrown_object</a>);</td></tr>
<tr><th id="620">620</th><td>        <b>if</b> (<a class="ref" href="#620" title='__sync_sub_and_fetch_8' data-ref="__sync_sub_and_fetch_8">__sync_sub_and_fetch</a>(&amp;<a class="local col7 ref" href="#67exception_header" title='exception_header' data-ref="67exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::referenceCount" title='__cxxabiv1::__cxa_exception::referenceCount' data-ref="__cxxabiv1::__cxa_exception::referenceCount">referenceCount</a>, <span class='typedef' title='size_t' data-type='unsigned long' data-ref="size_t">size_t</span>(<var>1</var>)) == <var>0</var>)</td></tr>
<tr><th id="621">621</th><td>        {</td></tr>
<tr><th id="622">622</th><td>            <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> != <a class="local col7 ref" href="#67exception_header" title='exception_header' data-ref="67exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::exceptionDestructor" title='__cxxabiv1::__cxa_exception::exceptionDestructor' data-ref="__cxxabiv1::__cxa_exception::exceptionDestructor">exceptionDestructor</a>)</td></tr>
<tr><th id="623">623</th><td>                <a class="local col7 ref" href="#67exception_header" title='exception_header' data-ref="67exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::exceptionDestructor" title='__cxxabiv1::__cxa_exception::exceptionDestructor' data-ref="__cxxabiv1::__cxa_exception::exceptionDestructor">exceptionDestructor</a>(<a class="local col6 ref" href="#66thrown_object" title='thrown_object' data-ref="66thrown_object">thrown_object</a>);</td></tr>
<tr><th id="624">624</th><td>            <a class="ref" href="#__cxa_free_exception" title='__cxa_free_exception' data-ref="__cxa_free_exception">__cxa_free_exception</a>(<a class="local col6 ref" href="#66thrown_object" title='thrown_object' data-ref="66thrown_object">thrown_object</a>);</td></tr>
<tr><th id="625">625</th><td>        }</td></tr>
<tr><th id="626">626</th><td>    }</td></tr>
<tr><th id="627">627</th><td>}</td></tr>
<tr><th id="628">628</th><td></td></tr>
<tr><th id="629">629</th><td><i>/*</i></td></tr>
<tr><th id="630">630</th><td><i>    Returns a pointer to the thrown object (if any) at the top of the</i></td></tr>
<tr><th id="631">631</th><td><i>    caughtExceptions stack.  Atomically increment the exception's referenceCount.</i></td></tr>
<tr><th id="632">632</th><td><i>    If there is no such thrown object or if the thrown object is foreign,</i></td></tr>
<tr><th id="633">633</th><td><i>    returns null.</i></td></tr>
<tr><th id="634">634</th><td><i></i></td></tr>
<tr><th id="635">635</th><td><i>    We can use __cxa_get_globals_fast here to get the globals because if there have</i></td></tr>
<tr><th id="636">636</th><td><i>    been no exceptions thrown, ever, on this thread, we can return NULL without </i></td></tr>
<tr><th id="637">637</th><td><i>    the need to allocate the exception-handling globals.</i></td></tr>
<tr><th id="638">638</th><td><i>*/</i></td></tr>
<tr><th id="639">639</th><td><em>void</em>*</td></tr>
<tr><th id="640">640</th><td><dfn class="decl def" id="__cxa_current_primary_exception" title='__cxa_current_primary_exception' data-ref="__cxa_current_primary_exception">__cxa_current_primary_exception</dfn>() <b>throw</b>()</td></tr>
<tr><th id="641">641</th><td>{</td></tr>
<tr><th id="642">642</th><td><i>//  get the current exception</i></td></tr>
<tr><th id="643">643</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals" title='__cxxabiv1::__cxa_eh_globals' data-ref="__cxxabiv1::__cxa_eh_globals">__cxa_eh_globals</a>* <dfn class="local col8 decl" id="68globals" title='globals' data-type='__cxxabiv1::__cxa_eh_globals *' data-ref="68globals">globals</dfn> = <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals_fast" title='__cxa_get_globals_fast' data-ref="__cxa_get_globals_fast">__cxa_get_globals_fast</a>();</td></tr>
<tr><th id="644">644</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col8 ref" href="#68globals" title='globals' data-ref="68globals">globals</a>)</td></tr>
<tr><th id="645">645</th><td>        <b>return</b> <span class="macro" title="__null" data-ref="_M/NULL">NULL</span>;        <i>//  If there are no globals, there is no exception</i></td></tr>
<tr><th id="646">646</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col9 decl" id="69exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="69exception_header">exception_header</dfn> = <a class="local col8 ref" href="#68globals" title='globals' data-ref="68globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::caughtExceptions" title='__cxxabiv1::__cxa_eh_globals::caughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::caughtExceptions">caughtExceptions</a>;</td></tr>
<tr><th id="647">647</th><td>    <b>if</b> (<span class="macro" title="__null" data-ref="_M/NULL">NULL</span> == <a class="local col9 ref" href="#69exception_header" title='exception_header' data-ref="69exception_header">exception_header</a>)</td></tr>
<tr><th id="648">648</th><td>        <b>return</b> <span class="macro" title="__null" data-ref="_M/NULL">NULL</span>;        <i>//  No current exception</i></td></tr>
<tr><th id="649">649</th><td>    <b>if</b> (!<a class="tu ref" href="#__cxxabiv1::isOurExceptionClass" title='__cxxabiv1::isOurExceptionClass' data-use='c' data-ref="__cxxabiv1::isOurExceptionClass">isOurExceptionClass</a>(&amp;<a class="local col9 ref" href="#69exception_header" title='exception_header' data-ref="69exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>))</td></tr>
<tr><th id="650">650</th><td>        <b>return</b> <span class="macro" title="__null" data-ref="_M/NULL">NULL</span>;        <i>// Can't capture a foreign exception (no way to refcount it)</i></td></tr>
<tr><th id="651">651</th><td>    <b>if</b> (<a class="tu ref" href="#__cxxabiv1::isDependentException" title='__cxxabiv1::isDependentException' data-use='c' data-ref="__cxxabiv1::isDependentException">isDependentException</a>(&amp;<a class="local col9 ref" href="#69exception_header" title='exception_header' data-ref="69exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::unwindHeader" title='__cxxabiv1::__cxa_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_exception::unwindHeader">unwindHeader</a>)) {</td></tr>
<tr><th id="652">652</th><td>        <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>* <dfn class="local col0 decl" id="70dep_exception_header" title='dep_exception_header' data-type='__cxxabiv1::__cxa_dependent_exception *' data-ref="70dep_exception_header">dep_exception_header</dfn> =</td></tr>
<tr><th id="653">653</th><td>            <b>reinterpret_cast</b>&lt;<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>*&gt;(<a class="local col9 ref" href="#69exception_header" title='exception_header' data-ref="69exception_header">exception_header</a>);</td></tr>
<tr><th id="654">654</th><td>        <a class="local col9 ref" href="#69exception_header" title='exception_header' data-ref="69exception_header">exception_header</a> = <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col0 ref" href="#70dep_exception_header" title='dep_exception_header' data-ref="70dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::primaryException" title='__cxxabiv1::__cxa_dependent_exception::primaryException' data-ref="__cxxabiv1::__cxa_dependent_exception::primaryException">primaryException</a>);</td></tr>
<tr><th id="655">655</th><td>    }</td></tr>
<tr><th id="656">656</th><td>    <em>void</em>* <dfn class="local col1 decl" id="71thrown_object" title='thrown_object' data-type='void *' data-ref="71thrown_object">thrown_object</dfn> = <a class="tu ref" href="#__cxxabiv1::thrown_object_from_cxa_exception" title='__cxxabiv1::thrown_object_from_cxa_exception' data-use='c' data-ref="__cxxabiv1::thrown_object_from_cxa_exception">thrown_object_from_cxa_exception</a>(<a class="local col9 ref" href="#69exception_header" title='exception_header' data-ref="69exception_header">exception_header</a>);</td></tr>
<tr><th id="657">657</th><td>    <a class="ref" href="#__cxa_increment_exception_refcount" title='__cxa_increment_exception_refcount' data-ref="__cxa_increment_exception_refcount">__cxa_increment_exception_refcount</a>(<a class="local col1 ref" href="#71thrown_object" title='thrown_object' data-ref="71thrown_object">thrown_object</a>);</td></tr>
<tr><th id="658">658</th><td>    <b>return</b> <a class="local col1 ref" href="#71thrown_object" title='thrown_object' data-ref="71thrown_object">thrown_object</a>;</td></tr>
<tr><th id="659">659</th><td>}</td></tr>
<tr><th id="660">660</th><td></td></tr>
<tr><th id="661">661</th><td><i  data-doc="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">/*</i></td></tr>
<tr><th id="662">662</th><td><i  data-doc="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">    If reason isn't _URC_FOREIGN_EXCEPTION_CAUGHT, then the terminateHandler</i></td></tr>
<tr><th id="663">663</th><td><i  data-doc="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">    stored in exc is called.  Otherwise the referenceCount stored in the</i></td></tr>
<tr><th id="664">664</th><td><i  data-doc="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">    primary exception is decremented, destroying the primary if necessary.</i></td></tr>
<tr><th id="665">665</th><td><i  data-doc="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">    Finally the dependent exception is destroyed.</i></td></tr>
<tr><th id="666">666</th><td><i  data-doc="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">*/</i></td></tr>
<tr><th id="667">667</th><td><em>static</em></td></tr>
<tr><th id="668">668</th><td><em>void</em></td></tr>
<tr><th id="669">669</th><td><dfn class="tu decl def" id="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception" title='dependent_exception_cleanup' data-type='void dependent_exception_cleanup(_Unwind_Reason_Code reason, _Unwind_Exception * unwind_exception)' data-ref="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">dependent_exception_cleanup</dfn>(<span class='typedef' title='_Unwind_Reason_Code' data-type='enum _Unwind_Reason_Code' data-ref="_Unwind_Reason_Code">_Unwind_Reason_Code</span> <dfn class="local col2 decl" id="72reason" title='reason' data-type='_Unwind_Reason_Code' data-ref="72reason">reason</dfn>, <span class='type' title='_Unwind_Exception' data-ref="_Unwind_Exception">_Unwind_Exception</span>* <dfn class="local col3 decl" id="73unwind_exception" title='unwind_exception' data-type='_Unwind_Exception *' data-ref="73unwind_exception">unwind_exception</dfn>)</td></tr>
<tr><th id="670">670</th><td>{</td></tr>
<tr><th id="671">671</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>* <dfn class="local col4 decl" id="74dep_exception_header" title='dep_exception_header' data-type='__cxxabiv1::__cxa_dependent_exception *' data-ref="74dep_exception_header">dep_exception_header</dfn> = </td></tr>
<tr><th id="672">672</th><td>                      <b>reinterpret_cast</b>&lt;<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>*&gt;(<a class="local col3 ref" href="#73unwind_exception" title='unwind_exception' data-ref="73unwind_exception">unwind_exception</a> + <var>1</var>) - <var>1</var>;</td></tr>
<tr><th id="673">673</th><td>    <b>if</b> (<span class='enum' title='_URC_FOREIGN_EXCEPTION_CAUGHT' data-ref="_URC_FOREIGN_EXCEPTION_CAUGHT">_URC_FOREIGN_EXCEPTION_CAUGHT</span> != <a class="local col2 ref" href="#72reason" title='reason' data-ref="72reason">reason</a>)</td></tr>
<tr><th id="674">674</th><td>        <span class="namespace">std::</span><a class="ref" href="cxa_handlers.hpp.html#_ZSt11__terminatePFvvE" title='std::__terminate' data-ref="_ZSt11__terminatePFvvE">__terminate</a>(<a class="local col4 ref" href="#74dep_exception_header" title='dep_exception_header' data-ref="74dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::terminateHandler" title='__cxxabiv1::__cxa_dependent_exception::terminateHandler' data-ref="__cxxabiv1::__cxa_dependent_exception::terminateHandler">terminateHandler</a>);</td></tr>
<tr><th id="675">675</th><td>    <a class="ref" href="#__cxa_decrement_exception_refcount" title='__cxa_decrement_exception_refcount' data-ref="__cxa_decrement_exception_refcount">__cxa_decrement_exception_refcount</a>(<a class="local col4 ref" href="#74dep_exception_header" title='dep_exception_header' data-ref="74dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::primaryException" title='__cxxabiv1::__cxa_dependent_exception::primaryException' data-ref="__cxxabiv1::__cxa_dependent_exception::primaryException">primaryException</a>);</td></tr>
<tr><th id="676">676</th><td>    <a class="ref" href="#__cxa_free_dependent_exception" title='__cxa_free_dependent_exception' data-ref="__cxa_free_dependent_exception">__cxa_free_dependent_exception</a>(<a class="local col4 ref" href="#74dep_exception_header" title='dep_exception_header' data-ref="74dep_exception_header">dep_exception_header</a>);</td></tr>
<tr><th id="677">677</th><td>}</td></tr>
<tr><th id="678">678</th><td></td></tr>
<tr><th id="679">679</th><td><i>/*</i></td></tr>
<tr><th id="680">680</th><td><i>    If thrown_object is not null, allocate, initialize and throw a dependent</i></td></tr>
<tr><th id="681">681</th><td><i>    exception.</i></td></tr>
<tr><th id="682">682</th><td><i>*/</i></td></tr>
<tr><th id="683">683</th><td><em>void</em></td></tr>
<tr><th id="684">684</th><td><dfn class="decl def" id="__cxa_rethrow_primary_exception" title='__cxa_rethrow_primary_exception' data-ref="__cxa_rethrow_primary_exception">__cxa_rethrow_primary_exception</dfn>(<em>void</em>* <dfn class="local col5 decl" id="75thrown_object" title='thrown_object' data-type='void *' data-ref="75thrown_object">thrown_object</dfn>)</td></tr>
<tr><th id="685">685</th><td>{</td></tr>
<tr><th id="686">686</th><td>    <b>if</b> ( <a class="local col5 ref" href="#75thrown_object" title='thrown_object' data-ref="75thrown_object">thrown_object</a> != <span class="macro" title="__null" data-ref="_M/NULL">NULL</span> )</td></tr>
<tr><th id="687">687</th><td>    {</td></tr>
<tr><th id="688">688</th><td>        <i>// thrown_object guaranteed to be native because</i></td></tr>
<tr><th id="689">689</th><td><i>        //   __cxa_current_primary_exception returns NULL for foreign exceptions</i></td></tr>
<tr><th id="690">690</th><td>        <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception" title='__cxxabiv1::__cxa_exception' data-ref="__cxxabiv1::__cxa_exception">__cxa_exception</a>* <dfn class="local col6 decl" id="76exception_header" title='exception_header' data-type='__cxxabiv1::__cxa_exception *' data-ref="76exception_header">exception_header</dfn> = <a class="tu ref" href="#__cxxabiv1::cxa_exception_from_thrown_object" title='__cxxabiv1::cxa_exception_from_thrown_object' data-use='c' data-ref="__cxxabiv1::cxa_exception_from_thrown_object">cxa_exception_from_thrown_object</a>(<a class="local col5 ref" href="#75thrown_object" title='thrown_object' data-ref="75thrown_object">thrown_object</a>);</td></tr>
<tr><th id="691">691</th><td>        <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>* <dfn class="local col7 decl" id="77dep_exception_header" title='dep_exception_header' data-type='__cxxabiv1::__cxa_dependent_exception *' data-ref="77dep_exception_header">dep_exception_header</dfn> =</td></tr>
<tr><th id="692">692</th><td>            <b>static_cast</b>&lt;<a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception" title='__cxxabiv1::__cxa_dependent_exception' data-ref="__cxxabiv1::__cxa_dependent_exception">__cxa_dependent_exception</a>*&gt;(<a class="ref" href="#__cxa_allocate_dependent_exception" title='__cxa_allocate_dependent_exception' data-ref="__cxa_allocate_dependent_exception">__cxa_allocate_dependent_exception</a>());</td></tr>
<tr><th id="693">693</th><td>        <a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::primaryException" title='__cxxabiv1::__cxa_dependent_exception::primaryException' data-ref="__cxxabiv1::__cxa_dependent_exception::primaryException">primaryException</a> = <a class="local col5 ref" href="#75thrown_object" title='thrown_object' data-ref="75thrown_object">thrown_object</a>;</td></tr>
<tr><th id="694">694</th><td>        <a class="ref" href="#__cxa_increment_exception_refcount" title='__cxa_increment_exception_refcount' data-ref="__cxa_increment_exception_refcount">__cxa_increment_exception_refcount</a>(<a class="local col5 ref" href="#75thrown_object" title='thrown_object' data-ref="75thrown_object">thrown_object</a>);</td></tr>
<tr><th id="695">695</th><td>        <a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::exceptionType" title='__cxxabiv1::__cxa_dependent_exception::exceptionType' data-ref="__cxxabiv1::__cxa_dependent_exception::exceptionType">exceptionType</a> = <a class="local col6 ref" href="#76exception_header" title='exception_header' data-ref="76exception_header">exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_exception::exceptionType" title='__cxxabiv1::__cxa_exception::exceptionType' data-ref="__cxxabiv1::__cxa_exception::exceptionType">exceptionType</a>;</td></tr>
<tr><th id="696">696</th><td>        <a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::unexpectedHandler" title='__cxxabiv1::__cxa_dependent_exception::unexpectedHandler' data-ref="__cxxabiv1::__cxa_dependent_exception::unexpectedHandler">unexpectedHandler</a> = <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt14get_unexpectedv" title='std::get_unexpected' data-ref="_ZSt14get_unexpectedv">get_unexpected</a>();</td></tr>
<tr><th id="697">697</th><td>        <a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::terminateHandler" title='__cxxabiv1::__cxa_dependent_exception::terminateHandler' data-ref="__cxxabiv1::__cxa_dependent_exception::terminateHandler">terminateHandler</a> = <span class="namespace">std::</span><a class="ref" href="../../libcxx/include/exception.html#_ZSt13get_terminatev" title='std::get_terminate' data-ref="_ZSt13get_terminatev">get_terminate</a>();</td></tr>
<tr><th id="698">698</th><td>        <a class="tu ref" href="#__cxxabiv1::setDependentExceptionClass" title='__cxxabiv1::setDependentExceptionClass' data-use='c' data-ref="__cxxabiv1::setDependentExceptionClass">setDependentExceptionClass</a>(&amp;<a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::unwindHeader" title='__cxxabiv1::__cxa_dependent_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_dependent_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="699">699</th><td>        <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals" title='__cxa_get_globals' data-ref="__cxa_get_globals">__cxa_get_globals</a>()-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::uncaughtExceptions" title='__cxxabiv1::__cxa_eh_globals::uncaughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::uncaughtExceptions">uncaughtExceptions</a> += <var>1</var>;</td></tr>
<tr><th id="700">700</th><td>        <a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::unwindHeader" title='__cxxabiv1::__cxa_dependent_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_dependent_exception::unwindHeader">unwindHeader</a>.<span class='ref' title='_Unwind_Exception::exception_cleanup' data-ref="_Unwind_Exception::exception_cleanup">exception_cleanup</span> = <a class="tu ref" href="#_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception" title='dependent_exception_cleanup' data-use='r' data-ref="_ZN10__cxxabiv127dependent_exception_cleanupE19_Unwind_Reason_CodeP17_Unwind_Exception">dependent_exception_cleanup</a>;</td></tr>
<tr><th id="701">701</th><td><u>#<span data-ppcond="701">ifdef</span> <span class="macro" data-ref="_M/__USING_SJLJ_EXCEPTIONS__">__USING_SJLJ_EXCEPTIONS__</span></u></td></tr>
<tr><th id="702">702</th><td>        _Unwind_SjLj_RaiseException(&amp;dep_exception_header-&gt;unwindHeader);</td></tr>
<tr><th id="703">703</th><td><u>#<span data-ppcond="701">else</span></u></td></tr>
<tr><th id="704">704</th><td>        <span class='ref' title='_Unwind_RaiseException' data-ref="_Unwind_RaiseException">_Unwind_RaiseException</span>(&amp;<a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::unwindHeader" title='__cxxabiv1::__cxa_dependent_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_dependent_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="705">705</th><td><u>#<span data-ppcond="701">endif</span></u></td></tr>
<tr><th id="706">706</th><td>        <i>// Some sort of unwinding error.  Note that terminate is a handler.</i></td></tr>
<tr><th id="707">707</th><td>        <a class="ref" href="#__cxa_begin_catch" title='__cxa_begin_catch' data-ref="__cxa_begin_catch">__cxa_begin_catch</a>(&amp;<a class="local col7 ref" href="#77dep_exception_header" title='dep_exception_header' data-ref="77dep_exception_header">dep_exception_header</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_dependent_exception::unwindHeader" title='__cxxabiv1::__cxa_dependent_exception::unwindHeader' data-ref="__cxxabiv1::__cxa_dependent_exception::unwindHeader">unwindHeader</a>);</td></tr>
<tr><th id="708">708</th><td>    }</td></tr>
<tr><th id="709">709</th><td>    <i>// If we return client will call terminate()</i></td></tr>
<tr><th id="710">710</th><td>}</td></tr>
<tr><th id="711">711</th><td></td></tr>
<tr><th id="712">712</th><td><em>bool</em></td></tr>
<tr><th id="713">713</th><td><dfn class="decl def" id="__cxa_uncaught_exception" title='__cxa_uncaught_exception' data-ref="__cxa_uncaught_exception">__cxa_uncaught_exception</dfn>() <b>throw</b>() { <b>return</b> <a class="ref" href="#__cxa_uncaught_exceptions" title='__cxa_uncaught_exceptions' data-ref="__cxa_uncaught_exceptions">__cxa_uncaught_exceptions</a>() != <var>0</var>; }</td></tr>
<tr><th id="714">714</th><td></td></tr>
<tr><th id="715">715</th><td><em>unsigned</em> <em>int</em></td></tr>
<tr><th id="716">716</th><td><dfn class="decl def" id="__cxa_uncaught_exceptions" title='__cxa_uncaught_exceptions' data-ref="__cxa_uncaught_exceptions">__cxa_uncaught_exceptions</dfn>() <b>throw</b>()</td></tr>
<tr><th id="717">717</th><td>{</td></tr>
<tr><th id="718">718</th><td>    <i>// This does not report foreign exceptions in flight</i></td></tr>
<tr><th id="719">719</th><td>    <a class="type" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals" title='__cxxabiv1::__cxa_eh_globals' data-ref="__cxxabiv1::__cxa_eh_globals">__cxa_eh_globals</a>* <dfn class="local col8 decl" id="78globals" title='globals' data-type='__cxxabiv1::__cxa_eh_globals *' data-ref="78globals">globals</dfn> = <a class="ref" href="cxa_exception.hpp.html#__cxa_get_globals_fast" title='__cxa_get_globals_fast' data-ref="__cxa_get_globals_fast">__cxa_get_globals_fast</a>();</td></tr>
<tr><th id="720">720</th><td>    <b>if</b> (<a class="local col8 ref" href="#78globals" title='globals' data-ref="78globals">globals</a> == <var>0</var>)</td></tr>
<tr><th id="721">721</th><td>        <b>return</b> <var>0</var>;</td></tr>
<tr><th id="722">722</th><td>    <b>return</b> <a class="local col8 ref" href="#78globals" title='globals' data-ref="78globals">globals</a>-&gt;<a class="ref" href="cxa_exception.hpp.html#__cxxabiv1::__cxa_eh_globals::uncaughtExceptions" title='__cxxabiv1::__cxa_eh_globals::uncaughtExceptions' data-ref="__cxxabiv1::__cxa_eh_globals::uncaughtExceptions">uncaughtExceptions</a>;</td></tr>
<tr><th id="723">723</th><td>}</td></tr>
<tr><th id="724">724</th><td></td></tr>
<tr><th id="725">725</th><td>}  <i>// extern "C"</i></td></tr>
<tr><th id="726">726</th><td></td></tr>
<tr><th id="727">727</th><td><u>#pragma GCC visibility pop</u></td></tr>
<tr><th id="728">728</th><td></td></tr>
<tr><th id="729">729</th><td>}  <i>// abi</i></td></tr>
<tr><th id="730">730</th><td></td></tr>
</table><hr/><p id='footer'>
Generated on <em>2017-Sep-15</em> from project codebrowser<br />Powered by <a href='https://woboq.com'><img alt='Woboq' src='https://code.woboq.org/woboq-16.png' width='41' height='16' /></a> <a href='https://code.woboq.org'>Code Browser</a> 2.1
<br/>Generator usage only permitted with license.</p>
</div></body></html>
